{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "cadee23e-1b03-4332-8d60-9c1239f30f8e",
   "metadata": {},
   "source": [
    "# 1. Parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "23e047ea-2e3f-4a42-867b-7112a179f261",
   "metadata": {},
   "outputs": [],
   "source": [
    "sequences_split_dir = 'input/split'\n",
    "sequences_subset_all = 'input/all.fasta.gz'\n",
    "pangolin_out_dir = 'input/pangolin'\n",
    "subset_metadata_path = 'input/metadata-subsample.tsv'\n",
    "output_metadata_pangolin_path = 'input/metadata-subsample-pangolin.tsv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ad50bedb-8447-4c3c-b407-f1812734aab7",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "\n",
    "sequences_split_dir = Path(sequences_split_dir)\n",
    "sequences_subset_all = Path(sequences_subset_all)\n",
    "pangolin_out_dir = Path(pangolin_out_dir)\n",
    "subset_metadata_path = Path(subset_metadata_path)\n",
    "output_metadata_pangolin_path = Path(output_metadata_pangolin_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "73ccbafc-81c6-41e9-962c-014ee21fd6bd",
   "metadata": {},
   "source": [
    "# 2. Run pangolin"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "e88ec771-2ba6-48a1-8d72-c1ba8d667dc3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pangolin: 3.1.11\n",
      "pangolearn: 2021-08-24\n",
      "constellations: v0.0.15\n",
      "scorpio: 0.3.12\n",
      "pango-designation: 1.2.76\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!conda run --name pangolin pangolin --all-versions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "fa9413c2-a261-4ade-9789-551f33132ba4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "loading model 09/07/2021, 19:07:13\n",
      "processing block of 1000 sequences 09/07/2021, 19:07:26\n",
      "processing block of 1000 sequences 09/07/2021, 19:07:51\n",
      "processing block of 1000 sequences 09/07/2021, 19:08:17\n",
      "processing block of 1000 sequences 09/07/2021, 19:08:42\n",
      "processing block of 1000 sequences 09/07/2021, 19:09:08\n",
      "processing block of 1000 sequences 09/07/2021, 19:09:34\n",
      "processing block of 1000 sequences 09/07/2021, 19:09:59\n",
      "processing block of 148 sequences 09/07/2021, 19:10:15\n",
      "complete 09/07/2021, 19:10:18\n",
      "\u001b[32mOutput file written to: \u001b[0m/home/CSCScience.ca/apetkau/workspace/genomics-data-index-evaluation/evaluations/sars-cov-2/input/pangolin/lineage_report.csv\n",
      "\u001b[32mAll dependencies satisfied.\u001b[0m\n",
      "\u001b[32mThe query file is:\u001b[0m/home/CSCScience.ca/apetkau/workspace/genomics-data-index-evaluation/evaluations/sars-cov-2/input/all.fasta.gz\n",
      "\u001b[32m** Sequence QC **\u001b[0m\n",
      "Sequence name                 \t                   Reason\tValue     \n",
      "\n",
      "OU419196                      \t       N content too high\t0.34      \n",
      "\n",
      "OU419643                      \t       N content too high\t0.32      \n",
      "\n",
      "LR991232                      \t       N content too high\t0.65      \n",
      "\n",
      "LR991192                      \t       N content too high\t0.85      \n",
      "\n",
      "USA/UT-UPHL-201007956/2020    \t       N content too high\t0.33      \n",
      "\n",
      "USA/UT-UPHL-201027561/2020    \t       N content too high\t0.32      \n",
      "\n",
      "USA/MA-Broad_UMMS-00174/2020  \t       N content too high\t0.46      \n",
      "\n",
      "Iceland/6451/2020             \t       N content too high\t0.62      \n",
      "\n",
      "LR963346                      \t       N content too high\t0.58      \n",
      "\n",
      "OU390314                      \t       N content too high\t0.38      \n",
      "\n",
      "Iceland/5021/2020             \t       N content too high\t0.4       \n",
      "\n",
      "Switzerland/BE-ETHZ-340169/2020\t       N content too high\t0.32      \n",
      "\n",
      "Switzerland/BE-ETHZ-350180/2020\t       N content too high\t0.32      \n",
      "\n",
      "USA/CA-UCSC-603/2020          \t       N content too high\t0.38      \n",
      "\n",
      "OU031475                      \t       N content too high\t0.42      \n",
      "\n",
      "USA/VA-CDC-LC0000688/2020     \t       N content too high\t0.36      \n",
      "\n",
      "USA/IN-CDC-LC0001462/2020     \t       N content too high\t0.39      \n",
      "\n",
      "USA/MD-CDC-LC0002143/2020     \t       N content too high\t0.35      \n",
      "\n",
      "USA/PA-CDC-LC0002911/2020     \t       N content too high\t0.33      \n",
      "\n",
      "USA/CA-CDC-LC0002796/2021     \t       N content too high\t0.43      \n",
      "\n",
      "USA/PA-CDC-LC0004486/2021     \t       N content too high\t0.39      \n",
      "\n",
      "USA/AR-CDC-LC0003058/2021     \t       N content too high\t0.32      \n",
      "\n",
      "USA/WI-CDC-LC0003284/2021     \t       N content too high\t0.46      \n",
      "\n",
      "USA/FL-CDC-LC0005920/2021     \t       N content too high\t0.36      \n",
      "\n",
      "USA/WV090175/2021             \t       N content too high\t0.36      \n",
      "\n",
      "USA/MA-MASPHL-01585/2021      \t       N content too high\t0.38      \n",
      "\n",
      "USA/VA-CDC-LC0007485/2021     \t       N content too high\t0.33      \n",
      "\n",
      "USA/MI-CDC-LC0008539/2021     \t       N content too high\t0.46      \n",
      "\n",
      "OU421574                      \t       N content too high\t0.37      \n",
      "\n",
      "USA/TN-CDC-LC0010167/2021     \t       N content too high\t0.33      \n",
      "\n",
      "USA/VA-CDC-LC0010506/2021     \t       N content too high\t0.36      \n",
      "\n",
      "USA/PA-CDC-LC0010952/2021     \t       N content too high\t0.48      \n",
      "\n",
      "USA/UND-00065852/2021         \t       N content too high\t0.49      \n",
      "\n",
      "USA/MA-CDC-LC0012142/2021     \t       N content too high\t0.4       \n",
      "\n",
      "USA/NJ-CDC-LC0011920/2021     \t       N content too high\t0.43      \n",
      "\n",
      "OU393363                      \t       N content too high\t0.42      \n",
      "\n",
      "USA/CT-CDC-LC0012626/2021     \t       N content too high\t0.49      \n",
      "\n",
      "USA/NM-CDC-LC0013284/2021     \t       N content too high\t0.32      \n",
      "\n",
      "USA/FL-CDC-LC0013765/2021     \t       N content too high\t0.44      \n",
      "\n",
      "OU029410                      \t       N content too high\t0.37      \n",
      "\n",
      "OU030431                      \t       N content too high\t0.34      \n",
      "\n",
      "USA/NC-CDC-LC0014313/2021     \t       N content too high\t0.32      \n",
      "\n",
      "USA/NJ-CDC-LC0014810/2021     \t       N content too high\t0.37      \n",
      "\n",
      "USA/NJ-CDC-LC0016358/2021     \t       N content too high\t0.43      \n",
      "\n",
      "USA/VA-CDC-LC0016116/2021     \t       N content too high\t0.34      \n",
      "\n",
      "USA/MD-CDC-LC0020341/2021     \t       N content too high\t0.36      \n",
      "\n",
      "USA/NC-CDC-LC0020386/2021     \t       N content too high\t0.4       \n",
      "\n",
      "USA/VA-CDC-LC0023369/2021     \t       N content too high\t0.47      \n",
      "\n",
      "USA/ID-CDC-LC0026419/2021     \t       N content too high\t0.33      \n",
      "\n",
      "USA/NJ-CDC-LC0026086/2021     \t       N content too high\t0.4       \n",
      "\n",
      "USA/CA-CDC-LC0027810/2021     \t       N content too high\t0.45      \n",
      "\n",
      "USA/NC-CDC-LC0025580/2021     \t       N content too high\t0.37      \n",
      "\n",
      "USA/NC-CDC-LC0026247/2021     \t       N content too high\t0.33      \n",
      "\n",
      "OU391494                      \t       N content too high\t0.43      \n",
      "\n",
      "USA/FL-CDC-LC0029864/2021     \t       N content too high\t0.33      \n",
      "\n",
      "OB998378                      \t       N content too high\t0.33      \n",
      "\n",
      "Estonia/Cov10570/2021         \t       N content too high\t0.44      \n",
      "\n",
      "Estonia/Cov11005/2021         \t       N content too high\t0.32      \n",
      "\n",
      "Switzerland/BE-ETHZ-610650/2021\t       N content too high\t0.31      \n",
      "\n",
      "OU420951                      \t       N content too high\t0.42      \n",
      "\n",
      "OU418445                      \t       N content too high\t0.44      \n",
      "\n",
      "OU347940                      \t       N content too high\t0.42      \n",
      "\n",
      "OU386014                      \t       N content too high\t0.36      \n",
      "\n",
      "OU447304                      \t       N content too high\t0.49      \n",
      "\n",
      "OU488487                      \t       N content too high\t0.36      \n",
      "\n",
      "HG999994                      \t       N content too high\t0.5       \n",
      "\n",
      "\u001b[32m\n",
      "Number of sequences detected: \u001b[0m10000\n",
      "\u001b[32mTotal passing QC: \u001b[0m9934\n",
      "\u001b[32m\n",
      "Data files found:\u001b[0m\n",
      "Trained model:\t/home/CSCScience.ca/apetkau/miniconda3/envs/pangolin/lib/python3.8/site-packages/pangoLEARN/data/decisionTree_v1.joblib\n",
      "Header file:\t/home/CSCScience.ca/apetkau/miniconda3/envs/pangolin/lib/python3.8/site-packages/pangoLEARN/data/decisionTreeHeaders_v1.joblib\n",
      "Designated hash:\t/home/CSCScience.ca/apetkau/miniconda3/envs/pangolin/lib/python3.8/site-packages/pangoLEARN/data/lineages.hash.csv\n",
      "\n",
      "Job stats:\n",
      "job                     count    min threads    max threads\n",
      "--------------------  -------  -------------  -------------\n",
      "add_failed_seqs             1              1              1\n",
      "align_to_reference          1              1              1\n",
      "all                         1              1              1\n",
      "generate_report             1              1              1\n",
      "get_constellations          1              1              1\n",
      "hash_sequence_assign        1              1              1\n",
      "pangolearn                  1              1              1\n",
      "scorpio                     1             52             52\n",
      "total                       8              1             52\n",
      "\n",
      "/home/CSCScience.ca/apetkau/miniconda3/envs/pangolin/lib/python3.8/site-packages/sklearn/base.py:329: UserWarning: Trying to unpickle estimator DecisionTreeClassifier from version 0.24.2 when using version 0.23.1. This might lead to breaking code or invalid results. Use at your own risk.\n",
      "  warnings.warn(\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!conda run --name pangolin pangolin -t 52 -o {pangolin_out_dir} {sequences_subset_all}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4897c5b3-bd6c-445f-a156-153ad3b97120",
   "metadata": {},
   "source": [
    "# 3. Load up and join pangolin to metadata\n",
    "\n",
    "## 3.1. Load pangolin results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c2e59423-ee34-4303-bc55-f44a9bca8a63",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>taxon</th>\n",
       "      <th>lineage</th>\n",
       "      <th>conflict</th>\n",
       "      <th>ambiguity_score</th>\n",
       "      <th>scorpio_call</th>\n",
       "      <th>scorpio_support</th>\n",
       "      <th>scorpio_conflict</th>\n",
       "      <th>version</th>\n",
       "      <th>pangolin_version</th>\n",
       "      <th>pangoLEARN_version</th>\n",
       "      <th>pango_version</th>\n",
       "      <th>status</th>\n",
       "      <th>note</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Switzerland/100118/2020</td>\n",
       "      <td>B.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PANGO-v1.2.66</td>\n",
       "      <td>3.1.11</td>\n",
       "      <td>2021-08-24</td>\n",
       "      <td>v1.2.66</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td>Assigned from designation hash.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>LR884325</td>\n",
       "      <td>B.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PANGO-v1.2.66</td>\n",
       "      <td>3.1.11</td>\n",
       "      <td>2021-08-24</td>\n",
       "      <td>v1.2.66</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td>Assigned from designation hash.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>LR884332</td>\n",
       "      <td>B.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PANGO-v1.2.66</td>\n",
       "      <td>3.1.11</td>\n",
       "      <td>2021-08-24</td>\n",
       "      <td>v1.2.66</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td>Assigned from designation hash.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>LR898961</td>\n",
       "      <td>B.1.177.23</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PANGO-v1.2.66</td>\n",
       "      <td>3.1.11</td>\n",
       "      <td>2021-08-24</td>\n",
       "      <td>v1.2.66</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td>Assigned from designation hash.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AUS/VIC16054/2020</td>\n",
       "      <td>D.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PANGO-v1.2.66</td>\n",
       "      <td>3.1.11</td>\n",
       "      <td>2021-08-24</td>\n",
       "      <td>v1.2.66</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td>Assigned from designation hash.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9995</th>\n",
       "      <td>OU347940</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PANGO-v1.2.66</td>\n",
       "      <td>3.1.11</td>\n",
       "      <td>2021-08-24</td>\n",
       "      <td>v1.2.66</td>\n",
       "      <td>fail</td>\n",
       "      <td>N_content:0.42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9996</th>\n",
       "      <td>OU386014</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PANGO-v1.2.66</td>\n",
       "      <td>3.1.11</td>\n",
       "      <td>2021-08-24</td>\n",
       "      <td>v1.2.66</td>\n",
       "      <td>fail</td>\n",
       "      <td>N_content:0.36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9997</th>\n",
       "      <td>OU447304</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PANGO-v1.2.66</td>\n",
       "      <td>3.1.11</td>\n",
       "      <td>2021-08-24</td>\n",
       "      <td>v1.2.66</td>\n",
       "      <td>fail</td>\n",
       "      <td>N_content:0.49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9998</th>\n",
       "      <td>OU488487</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PANGO-v1.2.66</td>\n",
       "      <td>3.1.11</td>\n",
       "      <td>2021-08-24</td>\n",
       "      <td>v1.2.66</td>\n",
       "      <td>fail</td>\n",
       "      <td>N_content:0.36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9999</th>\n",
       "      <td>HG999994</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PANGO-v1.2.66</td>\n",
       "      <td>3.1.11</td>\n",
       "      <td>2021-08-24</td>\n",
       "      <td>v1.2.66</td>\n",
       "      <td>fail</td>\n",
       "      <td>N_content:0.5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10000 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        taxon     lineage  conflict  ambiguity_score  \\\n",
       "0     Switzerland/100118/2020         B.1       NaN              NaN   \n",
       "1                    LR884325         B.1       NaN              NaN   \n",
       "2                    LR884332         B.1       NaN              NaN   \n",
       "3                    LR898961  B.1.177.23       NaN              NaN   \n",
       "4           AUS/VIC16054/2020         D.2       NaN              NaN   \n",
       "...                       ...         ...       ...              ...   \n",
       "9995                 OU347940        None       NaN              NaN   \n",
       "9996                 OU386014        None       NaN              NaN   \n",
       "9997                 OU447304        None       NaN              NaN   \n",
       "9998                 OU488487        None       NaN              NaN   \n",
       "9999                 HG999994        None       NaN              NaN   \n",
       "\n",
       "     scorpio_call  scorpio_support  scorpio_conflict        version  \\\n",
       "0             NaN              NaN               NaN  PANGO-v1.2.66   \n",
       "1             NaN              NaN               NaN  PANGO-v1.2.66   \n",
       "2             NaN              NaN               NaN  PANGO-v1.2.66   \n",
       "3             NaN              NaN               NaN  PANGO-v1.2.66   \n",
       "4             NaN              NaN               NaN  PANGO-v1.2.66   \n",
       "...           ...              ...               ...            ...   \n",
       "9995          NaN              NaN               NaN  PANGO-v1.2.66   \n",
       "9996          NaN              NaN               NaN  PANGO-v1.2.66   \n",
       "9997          NaN              NaN               NaN  PANGO-v1.2.66   \n",
       "9998          NaN              NaN               NaN  PANGO-v1.2.66   \n",
       "9999          NaN              NaN               NaN  PANGO-v1.2.66   \n",
       "\n",
       "     pangolin_version pangoLEARN_version pango_version     status  \\\n",
       "0              3.1.11         2021-08-24       v1.2.66  passed_qc   \n",
       "1              3.1.11         2021-08-24       v1.2.66  passed_qc   \n",
       "2              3.1.11         2021-08-24       v1.2.66  passed_qc   \n",
       "3              3.1.11         2021-08-24       v1.2.66  passed_qc   \n",
       "4              3.1.11         2021-08-24       v1.2.66  passed_qc   \n",
       "...               ...                ...           ...        ...   \n",
       "9995           3.1.11         2021-08-24       v1.2.66       fail   \n",
       "9996           3.1.11         2021-08-24       v1.2.66       fail   \n",
       "9997           3.1.11         2021-08-24       v1.2.66       fail   \n",
       "9998           3.1.11         2021-08-24       v1.2.66       fail   \n",
       "9999           3.1.11         2021-08-24       v1.2.66       fail   \n",
       "\n",
       "                                 note  \n",
       "0     Assigned from designation hash.  \n",
       "1     Assigned from designation hash.  \n",
       "2     Assigned from designation hash.  \n",
       "3     Assigned from designation hash.  \n",
       "4     Assigned from designation hash.  \n",
       "...                               ...  \n",
       "9995                   N_content:0.42  \n",
       "9996                   N_content:0.36  \n",
       "9997                   N_content:0.49  \n",
       "9998                   N_content:0.36  \n",
       "9999                    N_content:0.5  \n",
       "\n",
       "[10000 rows x 13 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "pangolin_report = pangolin_out_dir / 'lineage_report.csv'\n",
    "\n",
    "p_df = pd.read_csv(pangolin_report)\n",
    "p_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d199cae3-aa10-4d48-948e-209a2b38f726",
   "metadata": {},
   "source": [
    "## 3.2. Load other metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "bacd0e35-2d5d-48d1-90ed-329a05700367",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>strain</th>\n",
       "      <th>virus</th>\n",
       "      <th>gisaid_epi_isl</th>\n",
       "      <th>genbank_accession</th>\n",
       "      <th>sra_accession</th>\n",
       "      <th>date</th>\n",
       "      <th>region</th>\n",
       "      <th>country</th>\n",
       "      <th>division</th>\n",
       "      <th>location</th>\n",
       "      <th>...</th>\n",
       "      <th>divergence</th>\n",
       "      <th>nonACGTN</th>\n",
       "      <th>rare_mutations</th>\n",
       "      <th>snp_clusters</th>\n",
       "      <th>QC_missing_data</th>\n",
       "      <th>QC_mixed_sites</th>\n",
       "      <th>QC_rare_mutations</th>\n",
       "      <th>QC_snp_clusters</th>\n",
       "      <th>clock_deviation</th>\n",
       "      <th>missing_data_percent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>England/MILK-BB1C0F/2020</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_720301</td>\n",
       "      <td>OD907972</td>\n",
       "      <td>ERR5021413</td>\n",
       "      <td>2020-11-18</td>\n",
       "      <td>Europe</td>\n",
       "      <td>United Kingdom</td>\n",
       "      <td>England</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>19.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.404642</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>USA/MN-MDH-6083/2021</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_1820970</td>\n",
       "      <td>MZ066570</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2021-04-15</td>\n",
       "      <td>North America</td>\n",
       "      <td>USA</td>\n",
       "      <td>Minnesota</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>33.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>mediocre</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>-3.0</td>\n",
       "      <td>4.340702</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>USA/PA-CDC-STM-000058747/2021</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_1923233</td>\n",
       "      <td>MZ132284</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2021-04-18</td>\n",
       "      <td>North America</td>\n",
       "      <td>USA</td>\n",
       "      <td>Pennsylvania</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>43.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>USA/NJ-CDC-LC0045907/2021</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_1799871</td>\n",
       "      <td>MZ070532</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2021-04-09</td>\n",
       "      <td>North America</td>\n",
       "      <td>USA</td>\n",
       "      <td>New Jersey</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>20.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>-4.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>OU098600</td>\n",
       "      <td>ncov</td>\n",
       "      <td>?</td>\n",
       "      <td>OU098600</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Europe</td>\n",
       "      <td>Germany</td>\n",
       "      <td>North Rhine Westphalia</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>35.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>?</td>\n",
       "      <td>0.013377</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9995</th>\n",
       "      <td>England/MILK-19AF720/2021</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_3107867</td>\n",
       "      <td>OU464432</td>\n",
       "      <td>ERR6380849</td>\n",
       "      <td>2021-07-17</td>\n",
       "      <td>Europe</td>\n",
       "      <td>United Kingdom</td>\n",
       "      <td>England</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>33.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>mediocre</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>-4.0</td>\n",
       "      <td>3.905963</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9996</th>\n",
       "      <td>England/PLYM-19DF208/2021</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_3124091</td>\n",
       "      <td>OU467577</td>\n",
       "      <td>ERR6387813</td>\n",
       "      <td>2021-07-22</td>\n",
       "      <td>Europe</td>\n",
       "      <td>United Kingdom</td>\n",
       "      <td>England</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>38.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.404642</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9997</th>\n",
       "      <td>USA/CO-CDC-MMB08815626/2021</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_2440926</td>\n",
       "      <td>MZ354711</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2021-05-14</td>\n",
       "      <td>North America</td>\n",
       "      <td>USA</td>\n",
       "      <td>Colorado</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>44.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9998</th>\n",
       "      <td>England/ALDP-12C3E19/2021</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_1071782</td>\n",
       "      <td>OU011022</td>\n",
       "      <td>ERR5380776</td>\n",
       "      <td>2021-02-14</td>\n",
       "      <td>Europe</td>\n",
       "      <td>United Kingdom</td>\n",
       "      <td>England</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>32.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.404642</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9999</th>\n",
       "      <td>England/QEUH-17FE9C4/2021</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_2830208</td>\n",
       "      <td>OU392428</td>\n",
       "      <td>ERR6200982</td>\n",
       "      <td>2021-06-29</td>\n",
       "      <td>Europe</td>\n",
       "      <td>United Kingdom</td>\n",
       "      <td>England</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>35.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1.320938</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10000 rows × 40 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                             strain virus   gisaid_epi_isl genbank_accession  \\\n",
       "0          England/MILK-BB1C0F/2020  ncov   EPI_ISL_720301          OD907972   \n",
       "1              USA/MN-MDH-6083/2021  ncov  EPI_ISL_1820970          MZ066570   \n",
       "2     USA/PA-CDC-STM-000058747/2021  ncov  EPI_ISL_1923233          MZ132284   \n",
       "3         USA/NJ-CDC-LC0045907/2021  ncov  EPI_ISL_1799871          MZ070532   \n",
       "4                          OU098600  ncov                ?          OU098600   \n",
       "...                             ...   ...              ...               ...   \n",
       "9995      England/MILK-19AF720/2021  ncov  EPI_ISL_3107867          OU464432   \n",
       "9996      England/PLYM-19DF208/2021  ncov  EPI_ISL_3124091          OU467577   \n",
       "9997    USA/CO-CDC-MMB08815626/2021  ncov  EPI_ISL_2440926          MZ354711   \n",
       "9998      England/ALDP-12C3E19/2021  ncov  EPI_ISL_1071782          OU011022   \n",
       "9999      England/QEUH-17FE9C4/2021  ncov  EPI_ISL_2830208          OU392428   \n",
       "\n",
       "     sra_accession        date         region         country  \\\n",
       "0       ERR5021413  2020-11-18         Europe  United Kingdom   \n",
       "1              NaN  2021-04-15  North America             USA   \n",
       "2              NaN  2021-04-18  North America             USA   \n",
       "3              NaN  2021-04-09  North America             USA   \n",
       "4              NaN         NaN         Europe         Germany   \n",
       "...            ...         ...            ...             ...   \n",
       "9995    ERR6380849  2021-07-17         Europe  United Kingdom   \n",
       "9996    ERR6387813  2021-07-22         Europe  United Kingdom   \n",
       "9997           NaN  2021-05-14  North America             USA   \n",
       "9998    ERR5380776  2021-02-14         Europe  United Kingdom   \n",
       "9999    ERR6200982  2021-06-29         Europe  United Kingdom   \n",
       "\n",
       "                    division location  ... divergence nonACGTN rare_mutations  \\\n",
       "0                    England      NaN  ...       19.0      0.0            5.0   \n",
       "1                  Minnesota      NaN  ...       33.0      0.0            5.0   \n",
       "2               Pennsylvania      NaN  ...       43.0      0.0           15.0   \n",
       "3                 New Jersey      NaN  ...       20.0      0.0            4.0   \n",
       "4     North Rhine Westphalia      NaN  ...       35.0      0.0            8.0   \n",
       "...                      ...      ...  ...        ...      ...            ...   \n",
       "9995                 England      NaN  ...       33.0      0.0            6.0   \n",
       "9996                 England      NaN  ...       38.0      0.0            7.0   \n",
       "9997                Colorado      NaN  ...       44.0      0.0           12.0   \n",
       "9998                 England      NaN  ...       32.0      0.0            8.0   \n",
       "9999                 England      NaN  ...       35.0      0.0            6.0   \n",
       "\n",
       "     snp_clusters  QC_missing_data QC_mixed_sites QC_rare_mutations  \\\n",
       "0             0.0             good           good              good   \n",
       "1             0.0         mediocre           good              good   \n",
       "2             0.0             good           good              good   \n",
       "3             0.0             good           good              good   \n",
       "4             0.0             good           good              good   \n",
       "...           ...              ...            ...               ...   \n",
       "9995          0.0         mediocre           good              good   \n",
       "9996          0.0             good           good              good   \n",
       "9997          0.0             good           good              good   \n",
       "9998          0.0             good           good              good   \n",
       "9999          0.0             good           good              good   \n",
       "\n",
       "     QC_snp_clusters clock_deviation missing_data_percent  \n",
       "0               good             0.0             0.404642  \n",
       "1               good            -3.0             4.340702  \n",
       "2               good             6.0             0.000000  \n",
       "3               good            -4.0             0.000000  \n",
       "4               good               ?             0.013377  \n",
       "...              ...             ...                  ...  \n",
       "9995            good            -4.0             3.905963  \n",
       "9996            good             0.0             0.404642  \n",
       "9997            good             5.0             0.000000  \n",
       "9998            good             0.0             0.404642  \n",
       "9999            good            -1.0             1.320938  \n",
       "\n",
       "[10000 rows x 40 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "metadata_df = pd.read_csv(subset_metadata_path, sep='\\t')\n",
    "metadata_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "7fc53361-b9da-4763-89ed-1ee8332f24d1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['strain', 'virus', 'gisaid_epi_isl', 'genbank_accession',\n",
       "       'sra_accession', 'date', 'region', 'country', 'division', 'location',\n",
       "       'region_exposure', 'country_exposure', 'division_exposure', 'segment',\n",
       "       'length', 'host', 'age', 'sex', 'Nextstrain_clade', 'pango_lineage',\n",
       "       'GISAID_clade', 'originating_lab', 'submitting_lab', 'authors', 'url',\n",
       "       'title', 'paper_url', 'date_submitted', 'sampling_strategy',\n",
       "       'missing_data', 'divergence', 'nonACGTN', 'rare_mutations',\n",
       "       'snp_clusters', 'QC_missing_data', 'QC_mixed_sites',\n",
       "       'QC_rare_mutations', 'QC_snp_clusters', 'clock_deviation',\n",
       "       'missing_data_percent'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata_df.columns"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "580ea56f-1a60-4382-8eb7-e7db9b1650eb",
   "metadata": {},
   "source": [
    "## 3.3. Join to pangolin results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "94a19e11-f3ad-4206-8845-e1299a018115",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>strain</th>\n",
       "      <th>virus</th>\n",
       "      <th>gisaid_epi_isl</th>\n",
       "      <th>genbank_accession</th>\n",
       "      <th>sra_accession</th>\n",
       "      <th>date</th>\n",
       "      <th>region</th>\n",
       "      <th>country</th>\n",
       "      <th>division</th>\n",
       "      <th>location</th>\n",
       "      <th>...</th>\n",
       "      <th>ambiguity_score</th>\n",
       "      <th>scorpio_call</th>\n",
       "      <th>scorpio_support</th>\n",
       "      <th>scorpio_conflict</th>\n",
       "      <th>version</th>\n",
       "      <th>pangolin_version</th>\n",
       "      <th>pangoLEARN_version</th>\n",
       "      <th>pango_version</th>\n",
       "      <th>status</th>\n",
       "      <th>note</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>England/MILK-BB1C0F/2020</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_720301</td>\n",
       "      <td>OD907972</td>\n",
       "      <td>ERR5021413</td>\n",
       "      <td>2020-11-18</td>\n",
       "      <td>Europe</td>\n",
       "      <td>United Kingdom</td>\n",
       "      <td>England</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PANGO-v1.2.66</td>\n",
       "      <td>3.1.11</td>\n",
       "      <td>2021-08-24</td>\n",
       "      <td>v1.2.66</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td>Assigned from designation hash.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>USA/MN-MDH-6083/2021</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_1820970</td>\n",
       "      <td>MZ066570</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2021-04-15</td>\n",
       "      <td>North America</td>\n",
       "      <td>USA</td>\n",
       "      <td>Minnesota</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.946374</td>\n",
       "      <td>Alpha (B.1.1.7-like)</td>\n",
       "      <td>0.9130</td>\n",
       "      <td>0.0435</td>\n",
       "      <td>PLEARN-v1.2.66</td>\n",
       "      <td>3.1.11</td>\n",
       "      <td>2021-08-24</td>\n",
       "      <td>v1.2.66</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td>scorpio call: Alt alleles 21; Ref alleles 1; A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>USA/PA-CDC-STM-000058747/2021</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_1923233</td>\n",
       "      <td>MZ132284</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2021-04-18</td>\n",
       "      <td>North America</td>\n",
       "      <td>USA</td>\n",
       "      <td>Pennsylvania</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>Alpha (B.1.1.7-like)</td>\n",
       "      <td>0.9565</td>\n",
       "      <td>0.0435</td>\n",
       "      <td>PLEARN-v1.2.66</td>\n",
       "      <td>3.1.11</td>\n",
       "      <td>2021-08-24</td>\n",
       "      <td>v1.2.66</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td>scorpio call: Alt alleles 22; Ref alleles 1; A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>USA/NJ-CDC-LC0045907/2021</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_1799871</td>\n",
       "      <td>MZ070532</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2021-04-09</td>\n",
       "      <td>North America</td>\n",
       "      <td>USA</td>\n",
       "      <td>New Jersey</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>Iota (B.1.526-like)</td>\n",
       "      <td>0.8261</td>\n",
       "      <td>0.1739</td>\n",
       "      <td>PLEARN-v1.2.66</td>\n",
       "      <td>3.1.11</td>\n",
       "      <td>2021-08-24</td>\n",
       "      <td>v1.2.66</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td>scorpio call: Alt alleles 19; Ref alleles 4; A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>OU098600</td>\n",
       "      <td>ncov</td>\n",
       "      <td>?</td>\n",
       "      <td>OU098600</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Europe</td>\n",
       "      <td>Germany</td>\n",
       "      <td>North Rhine Westphalia</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>Alpha (B.1.1.7-like)</td>\n",
       "      <td>0.9565</td>\n",
       "      <td>0.0435</td>\n",
       "      <td>PLEARN-v1.2.66</td>\n",
       "      <td>3.1.11</td>\n",
       "      <td>2021-08-24</td>\n",
       "      <td>v1.2.66</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td>scorpio call: Alt alleles 22; Ref alleles 1; A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9995</th>\n",
       "      <td>England/MILK-19AF720/2021</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_3107867</td>\n",
       "      <td>OU464432</td>\n",
       "      <td>ERR6380849</td>\n",
       "      <td>2021-07-17</td>\n",
       "      <td>Europe</td>\n",
       "      <td>United Kingdom</td>\n",
       "      <td>England</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.961325</td>\n",
       "      <td>Delta (B.1.617.2-like)</td>\n",
       "      <td>0.7692</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>PLEARN-v1.2.66</td>\n",
       "      <td>3.1.11</td>\n",
       "      <td>2021-08-24</td>\n",
       "      <td>v1.2.66</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td>scorpio call: Alt alleles 10; Ref alleles 0; A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9996</th>\n",
       "      <td>England/PLYM-19DF208/2021</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_3124091</td>\n",
       "      <td>OU467577</td>\n",
       "      <td>ERR6387813</td>\n",
       "      <td>2021-07-22</td>\n",
       "      <td>Europe</td>\n",
       "      <td>United Kingdom</td>\n",
       "      <td>England</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>Delta (B.1.617.2-like)</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>PLEARN-v1.2.66</td>\n",
       "      <td>3.1.11</td>\n",
       "      <td>2021-08-24</td>\n",
       "      <td>v1.2.66</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td>scorpio call: Alt alleles 13; Ref alleles 0; A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9997</th>\n",
       "      <td>USA/CO-CDC-MMB08815626/2021</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_2440926</td>\n",
       "      <td>MZ354711</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2021-05-14</td>\n",
       "      <td>North America</td>\n",
       "      <td>USA</td>\n",
       "      <td>Colorado</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>Gamma (P.1-like)</td>\n",
       "      <td>0.8750</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>PLEARN-v1.2.66</td>\n",
       "      <td>3.1.11</td>\n",
       "      <td>2021-08-24</td>\n",
       "      <td>v1.2.66</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td>scorpio call: Alt alleles 14; Ref alleles 0; A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9998</th>\n",
       "      <td>England/ALDP-12C3E19/2021</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_1071782</td>\n",
       "      <td>OU011022</td>\n",
       "      <td>ERR5380776</td>\n",
       "      <td>2021-02-14</td>\n",
       "      <td>Europe</td>\n",
       "      <td>United Kingdom</td>\n",
       "      <td>England</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>Alpha (B.1.1.7-like)</td>\n",
       "      <td>0.9565</td>\n",
       "      <td>0.0435</td>\n",
       "      <td>PLEARN-v1.2.66</td>\n",
       "      <td>3.1.11</td>\n",
       "      <td>2021-08-24</td>\n",
       "      <td>v1.2.66</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td>scorpio call: Alt alleles 22; Ref alleles 1; A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9999</th>\n",
       "      <td>England/QEUH-17FE9C4/2021</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_2830208</td>\n",
       "      <td>OU392428</td>\n",
       "      <td>ERR6200982</td>\n",
       "      <td>2021-06-29</td>\n",
       "      <td>Europe</td>\n",
       "      <td>United Kingdom</td>\n",
       "      <td>England</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.989724</td>\n",
       "      <td>Delta (B.1.617.2-like)</td>\n",
       "      <td>0.9231</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>PLEARN-v1.2.66</td>\n",
       "      <td>3.1.11</td>\n",
       "      <td>2021-08-24</td>\n",
       "      <td>v1.2.66</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td>scorpio call: Alt alleles 12; Ref alleles 0; A...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10000 rows × 53 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                             strain virus   gisaid_epi_isl genbank_accession  \\\n",
       "0          England/MILK-BB1C0F/2020  ncov   EPI_ISL_720301          OD907972   \n",
       "1              USA/MN-MDH-6083/2021  ncov  EPI_ISL_1820970          MZ066570   \n",
       "2     USA/PA-CDC-STM-000058747/2021  ncov  EPI_ISL_1923233          MZ132284   \n",
       "3         USA/NJ-CDC-LC0045907/2021  ncov  EPI_ISL_1799871          MZ070532   \n",
       "4                          OU098600  ncov                ?          OU098600   \n",
       "...                             ...   ...              ...               ...   \n",
       "9995      England/MILK-19AF720/2021  ncov  EPI_ISL_3107867          OU464432   \n",
       "9996      England/PLYM-19DF208/2021  ncov  EPI_ISL_3124091          OU467577   \n",
       "9997    USA/CO-CDC-MMB08815626/2021  ncov  EPI_ISL_2440926          MZ354711   \n",
       "9998      England/ALDP-12C3E19/2021  ncov  EPI_ISL_1071782          OU011022   \n",
       "9999      England/QEUH-17FE9C4/2021  ncov  EPI_ISL_2830208          OU392428   \n",
       "\n",
       "     sra_accession        date         region         country  \\\n",
       "0       ERR5021413  2020-11-18         Europe  United Kingdom   \n",
       "1              NaN  2021-04-15  North America             USA   \n",
       "2              NaN  2021-04-18  North America             USA   \n",
       "3              NaN  2021-04-09  North America             USA   \n",
       "4              NaN         NaN         Europe         Germany   \n",
       "...            ...         ...            ...             ...   \n",
       "9995    ERR6380849  2021-07-17         Europe  United Kingdom   \n",
       "9996    ERR6387813  2021-07-22         Europe  United Kingdom   \n",
       "9997           NaN  2021-05-14  North America             USA   \n",
       "9998    ERR5380776  2021-02-14         Europe  United Kingdom   \n",
       "9999    ERR6200982  2021-06-29         Europe  United Kingdom   \n",
       "\n",
       "                    division location  ... ambiguity_score  \\\n",
       "0                    England      NaN  ...             NaN   \n",
       "1                  Minnesota      NaN  ...        0.946374   \n",
       "2               Pennsylvania      NaN  ...        1.000000   \n",
       "3                 New Jersey      NaN  ...        1.000000   \n",
       "4     North Rhine Westphalia      NaN  ...        1.000000   \n",
       "...                      ...      ...  ...             ...   \n",
       "9995                 England      NaN  ...        0.961325   \n",
       "9996                 England      NaN  ...        1.000000   \n",
       "9997                Colorado      NaN  ...        1.000000   \n",
       "9998                 England      NaN  ...        1.000000   \n",
       "9999                 England      NaN  ...        0.989724   \n",
       "\n",
       "                scorpio_call scorpio_support scorpio_conflict         version  \\\n",
       "0                        NaN             NaN              NaN   PANGO-v1.2.66   \n",
       "1       Alpha (B.1.1.7-like)          0.9130           0.0435  PLEARN-v1.2.66   \n",
       "2       Alpha (B.1.1.7-like)          0.9565           0.0435  PLEARN-v1.2.66   \n",
       "3        Iota (B.1.526-like)          0.8261           0.1739  PLEARN-v1.2.66   \n",
       "4       Alpha (B.1.1.7-like)          0.9565           0.0435  PLEARN-v1.2.66   \n",
       "...                      ...             ...              ...             ...   \n",
       "9995  Delta (B.1.617.2-like)          0.7692           0.0000  PLEARN-v1.2.66   \n",
       "9996  Delta (B.1.617.2-like)          1.0000           0.0000  PLEARN-v1.2.66   \n",
       "9997        Gamma (P.1-like)          0.8750           0.0000  PLEARN-v1.2.66   \n",
       "9998    Alpha (B.1.1.7-like)          0.9565           0.0435  PLEARN-v1.2.66   \n",
       "9999  Delta (B.1.617.2-like)          0.9231           0.0000  PLEARN-v1.2.66   \n",
       "\n",
       "     pangolin_version pangoLEARN_version pango_version     status  \\\n",
       "0              3.1.11         2021-08-24       v1.2.66  passed_qc   \n",
       "1              3.1.11         2021-08-24       v1.2.66  passed_qc   \n",
       "2              3.1.11         2021-08-24       v1.2.66  passed_qc   \n",
       "3              3.1.11         2021-08-24       v1.2.66  passed_qc   \n",
       "4              3.1.11         2021-08-24       v1.2.66  passed_qc   \n",
       "...               ...                ...           ...        ...   \n",
       "9995           3.1.11         2021-08-24       v1.2.66  passed_qc   \n",
       "9996           3.1.11         2021-08-24       v1.2.66  passed_qc   \n",
       "9997           3.1.11         2021-08-24       v1.2.66  passed_qc   \n",
       "9998           3.1.11         2021-08-24       v1.2.66  passed_qc   \n",
       "9999           3.1.11         2021-08-24       v1.2.66  passed_qc   \n",
       "\n",
       "                                                   note  \n",
       "0                       Assigned from designation hash.  \n",
       "1     scorpio call: Alt alleles 21; Ref alleles 1; A...  \n",
       "2     scorpio call: Alt alleles 22; Ref alleles 1; A...  \n",
       "3     scorpio call: Alt alleles 19; Ref alleles 4; A...  \n",
       "4     scorpio call: Alt alleles 22; Ref alleles 1; A...  \n",
       "...                                                 ...  \n",
       "9995  scorpio call: Alt alleles 10; Ref alleles 0; A...  \n",
       "9996  scorpio call: Alt alleles 13; Ref alleles 0; A...  \n",
       "9997  scorpio call: Alt alleles 14; Ref alleles 0; A...  \n",
       "9998  scorpio call: Alt alleles 22; Ref alleles 1; A...  \n",
       "9999  scorpio call: Alt alleles 12; Ref alleles 0; A...  \n",
       "\n",
       "[10000 rows x 53 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata_pangolin_df = metadata_df.merge(p_df, left_on='strain', right_on='taxon')\n",
    "metadata_pangolin_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "58d82e66-28b1-4882-8b28-b422eacf3041",
   "metadata": {},
   "source": [
    "# 4. Save joined metadata to file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "9761daae-e204-42d7-84fc-9e49b775fde7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saved metadata and pangolin to input/metadata-subsample-pangolin.tsv\n"
     ]
    }
   ],
   "source": [
    "metadata_pangolin_df.to_csv(output_metadata_pangolin_path, sep='\\t', index=False)\n",
    "\n",
    "print(f'Saved metadata and pangolin to {output_metadata_pangolin_path}')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
