{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "cadee23e-1b03-4332-8d60-9c1239f30f8e",
   "metadata": {},
   "source": [
    "# 1. Parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "23e047ea-2e3f-4a42-867b-7112a179f261",
   "metadata": {},
   "outputs": [],
   "source": [
    "sequences_split_dir = 'input/split'\n",
    "sequences_subset_all = 'input/all.fasta.gz'\n",
    "pangolin_out_dir = 'input/pangolin'\n",
    "subset_metadata_path = 'input/metadata-subsample.tsv'\n",
    "output_metadata_pangolin_path = 'input/metadata-subsample-pangolin.tsv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ad50bedb-8447-4c3c-b407-f1812734aab7",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "\n",
    "sequences_split_dir = Path(sequences_split_dir)\n",
    "sequences_subset_all = Path(sequences_subset_all)\n",
    "pangolin_out_dir = Path(pangolin_out_dir)\n",
    "subset_metadata_path = Path(subset_metadata_path)\n",
    "output_metadata_pangolin_path = Path(output_metadata_pangolin_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "73ccbafc-81c6-41e9-962c-014ee21fd6bd",
   "metadata": {},
   "source": [
    "# 2. Run pangolin"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "e88ec771-2ba6-48a1-8d72-c1ba8d667dc3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pangolin: 3.1.20\n",
      "pangolearn: 2022-02-28\n",
      "constellations: v0.1.3\n",
      "scorpio: 0.3.16\n",
      "pango-designation used by pangoLEARN/Usher: v1.2.127\n",
      "pango-designation aliases: 1.2.132\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!conda run --name pangolin pangolin --all-versions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "fa9413c2-a261-4ade-9789-551f33132ba4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "loading model 03/10/2022, 11:05:34\n",
      "processing block of 1000 sequences 03/10/2022, 11:05:42\n",
      "processing block of 1000 sequences 03/10/2022, 11:05:47\n",
      "processing block of 1000 sequences 03/10/2022, 11:05:52\n",
      "processing block of 1000 sequences 03/10/2022, 11:05:56\n",
      "processing block of 1000 sequences 03/10/2022, 11:06:01\n",
      "processing block of 1000 sequences 03/10/2022, 11:06:06\n",
      "processing block of 1000 sequences 03/10/2022, 11:06:11\n",
      "processing block of 1000 sequences 03/10/2022, 11:06:15\n",
      "processing block of 1000 sequences 03/10/2022, 11:06:20\n",
      "processing block of 1000 sequences 03/10/2022, 11:06:25\n",
      "processing block of 1000 sequences 03/10/2022, 11:06:30\n",
      "processing block of 1000 sequences 03/10/2022, 11:06:34\n",
      "processing block of 1000 sequences 03/10/2022, 11:06:39\n",
      "processing block of 1000 sequences 03/10/2022, 11:06:44\n",
      "processing block of 1000 sequences 03/10/2022, 11:06:48\n",
      "processing block of 1000 sequences 03/10/2022, 11:06:53\n",
      "processing block of 1000 sequences 03/10/2022, 11:06:58\n",
      "processing block of 1000 sequences 03/10/2022, 11:07:03\n",
      "processing block of 1000 sequences 03/10/2022, 11:07:07\n",
      "processing block of 1000 sequences 03/10/2022, 11:07:12\n",
      "processing block of 1000 sequences 03/10/2022, 11:07:17\n",
      "processing block of 1000 sequences 03/10/2022, 11:07:21\n",
      "processing block of 1000 sequences 03/10/2022, 11:07:26\n",
      "processing block of 1000 sequences 03/10/2022, 11:07:31\n",
      "processing block of 1000 sequences 03/10/2022, 11:07:36\n",
      "processing block of 1000 sequences 03/10/2022, 11:07:40\n",
      "processing block of 1000 sequences 03/10/2022, 11:07:45\n",
      "processing block of 1000 sequences 03/10/2022, 11:07:50\n",
      "processing block of 1000 sequences 03/10/2022, 11:07:55\n",
      "processing block of 1000 sequences 03/10/2022, 11:07:59\n",
      "processing block of 1000 sequences 03/10/2022, 11:08:04\n",
      "processing block of 1000 sequences 03/10/2022, 11:08:09\n",
      "processing block of 1000 sequences 03/10/2022, 11:08:14\n",
      "processing block of 1000 sequences 03/10/2022, 11:08:18\n",
      "processing block of 1000 sequences 03/10/2022, 11:08:23\n",
      "processing block of 1000 sequences 03/10/2022, 11:08:28\n",
      "processing block of 1000 sequences 03/10/2022, 11:08:33\n",
      "processing block of 1000 sequences 03/10/2022, 11:08:38\n",
      "processing block of 1000 sequences 03/10/2022, 11:08:42\n",
      "processing block of 1000 sequences 03/10/2022, 11:08:47\n",
      "processing block of 1000 sequences 03/10/2022, 11:08:52\n",
      "processing block of 1000 sequences 03/10/2022, 11:08:57\n",
      "processing block of 1000 sequences 03/10/2022, 11:09:02\n",
      "processing block of 1000 sequences 03/10/2022, 11:09:07\n",
      "processing block of 1000 sequences 03/10/2022, 11:09:12\n",
      "processing block of 1000 sequences 03/10/2022, 11:09:17\n",
      "processing block of 1000 sequences 03/10/2022, 11:09:21\n",
      "processing block of 1000 sequences 03/10/2022, 11:09:26\n",
      "processing block of 1000 sequences 03/10/2022, 11:09:31\n",
      "processing block of 1000 sequences 03/10/2022, 11:09:36\n",
      "processing block of 1000 sequences 03/10/2022, 11:09:40\n",
      "processing block of 1000 sequences 03/10/2022, 11:09:45\n",
      "processing block of 1000 sequences 03/10/2022, 11:09:50\n",
      "processing block of 1000 sequences 03/10/2022, 11:09:55\n",
      "processing block of 1000 sequences 03/10/2022, 11:09:59\n",
      "processing block of 1000 sequences 03/10/2022, 11:10:04\n",
      "processing block of 1000 sequences 03/10/2022, 11:10:09\n",
      "processing block of 1000 sequences 03/10/2022, 11:10:13\n",
      "processing block of 1000 sequences 03/10/2022, 11:10:18\n",
      "processing block of 1000 sequences 03/10/2022, 11:10:23\n",
      "processing block of 1000 sequences 03/10/2022, 11:10:28\n",
      "processing block of 1000 sequences 03/10/2022, 11:10:32\n",
      "processing block of 1000 sequences 03/10/2022, 11:10:37\n",
      "processing block of 1000 sequences 03/10/2022, 11:10:42\n",
      "processing block of 1000 sequences 03/10/2022, 11:10:46\n",
      "processing block of 1000 sequences 03/10/2022, 11:10:51\n",
      "processing block of 1000 sequences 03/10/2022, 11:10:56\n",
      "processing block of 1000 sequences 03/10/2022, 11:11:01\n",
      "processing block of 1000 sequences 03/10/2022, 11:11:05\n",
      "processing block of 441 sequences 03/10/2022, 11:11:09\n",
      "complete 03/10/2022, 11:11:10\n",
      "\u001b[32mOutput file written to: \u001b[0m/home/CSCScience.ca/apetkau/workspace/genomics-data-index-evaluation/evaluations/sars-cov-2/input/pangolin/lineage_report.csv\n",
      "\u001b[32mAll dependencies satisfied.\u001b[0m\n",
      "\u001b[32mThe query file is:\u001b[0m/home/CSCScience.ca/apetkau/workspace/genomics-data-index-evaluation/evaluations/sars-cov-2/input/all.fasta.gz\n",
      "\u001b[32m** Running sequence QC **\u001b[0m\n",
      "\u001b[32mNumber of sequences detected: \u001b[0m100000\n",
      "\u001b[32mTotal passing QC: \u001b[0m99492\n",
      "\u001b[32m\n",
      "Data files found:\u001b[0m\n",
      "Trained model:\t/home/CSCScience.ca/apetkau/miniconda3/envs/pangolin/lib/python3.8/site-packages/pangoLEARN/data/decisionTree_v1.joblib\n",
      "Header file:\t/home/CSCScience.ca/apetkau/miniconda3/envs/pangolin/lib/python3.8/site-packages/pangoLEARN/data/decisionTreeHeaders_v1.joblib\n",
      "Designated hash:\t/home/CSCScience.ca/apetkau/miniconda3/envs/pangolin/lib/python3.8/site-packages/pangoLEARN/data/lineages.hash.csv\n",
      "\n",
      "Job stats:\n",
      "job                     count    min threads    max threads\n",
      "--------------------  -------  -------------  -------------\n",
      "add_failed_seqs             1              1              1\n",
      "align_to_reference          1              1              1\n",
      "all                         1              1              1\n",
      "generate_report             1              1              1\n",
      "get_constellations          1              1              1\n",
      "hash_sequence_assign        1              1              1\n",
      "pangolearn                  1              1              1\n",
      "scorpio                     1             52             52\n",
      "total                       8              1             52\n",
      "\n",
      "/home/CSCScience.ca/apetkau/miniconda3/envs/pangolin/lib/python3.8/site-packages/sklearn/base.py:329: UserWarning: Trying to unpickle estimator DecisionTreeClassifier from version 0.24.2 when using version 0.23.1. This might lead to breaking code or invalid results. Use at your own risk.\n",
      "  warnings.warn(\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!conda run --name pangolin pangolin -t 52 -o {pangolin_out_dir} {sequences_subset_all}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4897c5b3-bd6c-445f-a156-153ad3b97120",
   "metadata": {},
   "source": [
    "# 3. Load up and join pangolin to metadata\n",
    "\n",
    "## 3.1. Load pangolin results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c2e59423-ee34-4303-bc55-f44a9bca8a63",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>taxon</th>\n",
       "      <th>lineage</th>\n",
       "      <th>conflict</th>\n",
       "      <th>ambiguity_score</th>\n",
       "      <th>scorpio_call</th>\n",
       "      <th>scorpio_support</th>\n",
       "      <th>scorpio_conflict</th>\n",
       "      <th>version</th>\n",
       "      <th>pangolin_version</th>\n",
       "      <th>pangoLEARN_version</th>\n",
       "      <th>pango_version</th>\n",
       "      <th>status</th>\n",
       "      <th>note</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Wuhan/IPBCAMS-WH-04/2019</td>\n",
       "      <td>B</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PANGO-v1.2.127</td>\n",
       "      <td>3.1.20</td>\n",
       "      <td>2022-02-28</td>\n",
       "      <td>v1.2.127</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td>Assigned from designation hash.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHN/Wuhan_IME-WH05/2019</td>\n",
       "      <td>B</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PANGO-v1.2.127</td>\n",
       "      <td>3.1.20</td>\n",
       "      <td>2022-02-28</td>\n",
       "      <td>v1.2.127</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td>Assigned from designation hash.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Switzerland/100108/2020</td>\n",
       "      <td>B.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PANGO-v1.2.127</td>\n",
       "      <td>3.1.20</td>\n",
       "      <td>2022-02-28</td>\n",
       "      <td>v1.2.127</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td>Assigned from designation hash.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Switzerland/100816/2020</td>\n",
       "      <td>B.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PANGO-v1.2.127</td>\n",
       "      <td>3.1.20</td>\n",
       "      <td>2022-02-28</td>\n",
       "      <td>v1.2.127</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td>Assigned from designation hash.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LR862442</td>\n",
       "      <td>C.35</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PANGO-v1.2.127</td>\n",
       "      <td>3.1.20</td>\n",
       "      <td>2022-02-28</td>\n",
       "      <td>v1.2.127</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td>Assigned from designation hash.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99995</th>\n",
       "      <td>OU488207</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PANGO-v1.2.127</td>\n",
       "      <td>3.1.20</td>\n",
       "      <td>2022-02-28</td>\n",
       "      <td>v1.2.127</td>\n",
       "      <td>fail</td>\n",
       "      <td>N_content:0.41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99996</th>\n",
       "      <td>OU490562</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PANGO-v1.2.127</td>\n",
       "      <td>3.1.20</td>\n",
       "      <td>2022-02-28</td>\n",
       "      <td>v1.2.127</td>\n",
       "      <td>fail</td>\n",
       "      <td>N_content:0.47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99997</th>\n",
       "      <td>OU490569</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PANGO-v1.2.127</td>\n",
       "      <td>3.1.20</td>\n",
       "      <td>2022-02-28</td>\n",
       "      <td>v1.2.127</td>\n",
       "      <td>fail</td>\n",
       "      <td>N_content:0.45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99998</th>\n",
       "      <td>OU490836</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PANGO-v1.2.127</td>\n",
       "      <td>3.1.20</td>\n",
       "      <td>2022-02-28</td>\n",
       "      <td>v1.2.127</td>\n",
       "      <td>fail</td>\n",
       "      <td>N_content:0.31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99999</th>\n",
       "      <td>USA/D4F17D14A7C44B168CD47F1B0D34ADB7/2021</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PANGO-v1.2.127</td>\n",
       "      <td>3.1.20</td>\n",
       "      <td>2022-02-28</td>\n",
       "      <td>v1.2.127</td>\n",
       "      <td>fail</td>\n",
       "      <td>N_content:0.32</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100000 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           taxon lineage  conflict  \\\n",
       "0                       Wuhan/IPBCAMS-WH-04/2019       B       NaN   \n",
       "1                        CHN/Wuhan_IME-WH05/2019       B       NaN   \n",
       "2                        Switzerland/100108/2020     B.1       NaN   \n",
       "3                        Switzerland/100816/2020     B.1       NaN   \n",
       "4                                       LR862442    C.35       NaN   \n",
       "...                                          ...     ...       ...   \n",
       "99995                                   OU488207    None       NaN   \n",
       "99996                                   OU490562    None       NaN   \n",
       "99997                                   OU490569    None       NaN   \n",
       "99998                                   OU490836    None       NaN   \n",
       "99999  USA/D4F17D14A7C44B168CD47F1B0D34ADB7/2021    None       NaN   \n",
       "\n",
       "       ambiguity_score scorpio_call  scorpio_support  scorpio_conflict  \\\n",
       "0                  NaN          NaN              NaN               NaN   \n",
       "1                  NaN          NaN              NaN               NaN   \n",
       "2                  NaN          NaN              NaN               NaN   \n",
       "3                  NaN          NaN              NaN               NaN   \n",
       "4                  NaN          NaN              NaN               NaN   \n",
       "...                ...          ...              ...               ...   \n",
       "99995              NaN          NaN              NaN               NaN   \n",
       "99996              NaN          NaN              NaN               NaN   \n",
       "99997              NaN          NaN              NaN               NaN   \n",
       "99998              NaN          NaN              NaN               NaN   \n",
       "99999              NaN          NaN              NaN               NaN   \n",
       "\n",
       "              version pangolin_version pangoLEARN_version pango_version  \\\n",
       "0      PANGO-v1.2.127           3.1.20         2022-02-28      v1.2.127   \n",
       "1      PANGO-v1.2.127           3.1.20         2022-02-28      v1.2.127   \n",
       "2      PANGO-v1.2.127           3.1.20         2022-02-28      v1.2.127   \n",
       "3      PANGO-v1.2.127           3.1.20         2022-02-28      v1.2.127   \n",
       "4      PANGO-v1.2.127           3.1.20         2022-02-28      v1.2.127   \n",
       "...               ...              ...                ...           ...   \n",
       "99995  PANGO-v1.2.127           3.1.20         2022-02-28      v1.2.127   \n",
       "99996  PANGO-v1.2.127           3.1.20         2022-02-28      v1.2.127   \n",
       "99997  PANGO-v1.2.127           3.1.20         2022-02-28      v1.2.127   \n",
       "99998  PANGO-v1.2.127           3.1.20         2022-02-28      v1.2.127   \n",
       "99999  PANGO-v1.2.127           3.1.20         2022-02-28      v1.2.127   \n",
       "\n",
       "          status                             note  \n",
       "0      passed_qc  Assigned from designation hash.  \n",
       "1      passed_qc  Assigned from designation hash.  \n",
       "2      passed_qc  Assigned from designation hash.  \n",
       "3      passed_qc  Assigned from designation hash.  \n",
       "4      passed_qc  Assigned from designation hash.  \n",
       "...          ...                              ...  \n",
       "99995       fail                   N_content:0.41  \n",
       "99996       fail                   N_content:0.47  \n",
       "99997       fail                   N_content:0.45  \n",
       "99998       fail                   N_content:0.31  \n",
       "99999       fail                   N_content:0.32  \n",
       "\n",
       "[100000 rows x 13 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "pangolin_report = pangolin_out_dir / 'lineage_report.csv'\n",
    "\n",
    "p_df = pd.read_csv(pangolin_report)\n",
    "p_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d199cae3-aa10-4d48-948e-209a2b38f726",
   "metadata": {},
   "source": [
    "## 3.2. Load other metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "bacd0e35-2d5d-48d1-90ed-329a05700367",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>strain</th>\n",
       "      <th>virus</th>\n",
       "      <th>gisaid_epi_isl</th>\n",
       "      <th>genbank_accession</th>\n",
       "      <th>sra_accession</th>\n",
       "      <th>date</th>\n",
       "      <th>region</th>\n",
       "      <th>country</th>\n",
       "      <th>division</th>\n",
       "      <th>location</th>\n",
       "      <th>...</th>\n",
       "      <th>divergence</th>\n",
       "      <th>nonACGTN</th>\n",
       "      <th>rare_mutations</th>\n",
       "      <th>snp_clusters</th>\n",
       "      <th>QC_missing_data</th>\n",
       "      <th>QC_mixed_sites</th>\n",
       "      <th>QC_rare_mutations</th>\n",
       "      <th>QC_snp_clusters</th>\n",
       "      <th>clock_deviation</th>\n",
       "      <th>missing_data_percent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>OU420663</td>\n",
       "      <td>ncov</td>\n",
       "      <td>?</td>\n",
       "      <td>OU420663</td>\n",
       "      <td>ERR5939958</td>\n",
       "      <td>2020</td>\n",
       "      <td>Europe</td>\n",
       "      <td>United Kingdom</td>\n",
       "      <td>England</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>40.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>?</td>\n",
       "      <td>0.351135</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>USA/NY-CUIMC-NP-3606/2020</td>\n",
       "      <td>ncov</td>\n",
       "      <td>?</td>\n",
       "      <td>MZ702266</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2020-12-02</td>\n",
       "      <td>North America</td>\n",
       "      <td>USA</td>\n",
       "      <td>New York</td>\n",
       "      <td>New York City</td>\n",
       "      <td>...</td>\n",
       "      <td>21.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>mediocre</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.417617</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Switzerland/BL-UHB-42491849/2020</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_930932</td>\n",
       "      <td>OU281100</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2020-10-18</td>\n",
       "      <td>Europe</td>\n",
       "      <td>Switzerland</td>\n",
       "      <td>Basel-Land</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>19.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.297629</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Scotland/QEUH-1618558/2021</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_2487678</td>\n",
       "      <td>OU314699</td>\n",
       "      <td>ERR6069165</td>\n",
       "      <td>2021-06-02</td>\n",
       "      <td>Europe</td>\n",
       "      <td>United Kingdom</td>\n",
       "      <td>Scotland</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>33.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>mediocre</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>-5.0</td>\n",
       "      <td>3.885898</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Scotland/QEUH-16829C6/2021</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_2554507</td>\n",
       "      <td>OU329920</td>\n",
       "      <td>ERR6085250</td>\n",
       "      <td>2021-06-10</td>\n",
       "      <td>Europe</td>\n",
       "      <td>United Kingdom</td>\n",
       "      <td>Scotland</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>37.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.404642</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99995</th>\n",
       "      <td>USA/OH-CDC-QDX23654955/2021</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_1669162</td>\n",
       "      <td>MW989219</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2021-04-05</td>\n",
       "      <td>North America</td>\n",
       "      <td>USA</td>\n",
       "      <td>Ohio</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>35.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99996</th>\n",
       "      <td>USA/KS-CDC-LC0014845/2021</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_1255089</td>\n",
       "      <td>MW704675</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2021-02-09</td>\n",
       "      <td>North America</td>\n",
       "      <td>USA</td>\n",
       "      <td>Kansas</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>21.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>-2.0</td>\n",
       "      <td>3.394308</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99997</th>\n",
       "      <td>USA/WA-S5065/2020</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_1114222</td>\n",
       "      <td>MW689132</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2020-11-06</td>\n",
       "      <td>North America</td>\n",
       "      <td>USA</td>\n",
       "      <td>Washington</td>\n",
       "      <td>Grays Harbor County</td>\n",
       "      <td>...</td>\n",
       "      <td>14.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>-2.0</td>\n",
       "      <td>0.006688</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99998</th>\n",
       "      <td>England/MILK-10AE881/2021</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_993691</td>\n",
       "      <td>OD978562</td>\n",
       "      <td>ERR5303908</td>\n",
       "      <td>2021-01-17</td>\n",
       "      <td>Europe</td>\n",
       "      <td>United Kingdom</td>\n",
       "      <td>England</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>32.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.595158</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99999</th>\n",
       "      <td>England/ALDP-B5B038/2020</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_655604</td>\n",
       "      <td>OB997017</td>\n",
       "      <td>ERR4877748</td>\n",
       "      <td>2020-11-08</td>\n",
       "      <td>Europe</td>\n",
       "      <td>United Kingdom</td>\n",
       "      <td>England</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>16.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>2.324182</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100000 rows × 40 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 strain virus   gisaid_epi_isl  \\\n",
       "0                              OU420663  ncov                ?   \n",
       "1             USA/NY-CUIMC-NP-3606/2020  ncov                ?   \n",
       "2      Switzerland/BL-UHB-42491849/2020  ncov   EPI_ISL_930932   \n",
       "3            Scotland/QEUH-1618558/2021  ncov  EPI_ISL_2487678   \n",
       "4            Scotland/QEUH-16829C6/2021  ncov  EPI_ISL_2554507   \n",
       "...                                 ...   ...              ...   \n",
       "99995       USA/OH-CDC-QDX23654955/2021  ncov  EPI_ISL_1669162   \n",
       "99996         USA/KS-CDC-LC0014845/2021  ncov  EPI_ISL_1255089   \n",
       "99997                 USA/WA-S5065/2020  ncov  EPI_ISL_1114222   \n",
       "99998         England/MILK-10AE881/2021  ncov   EPI_ISL_993691   \n",
       "99999          England/ALDP-B5B038/2020  ncov   EPI_ISL_655604   \n",
       "\n",
       "      genbank_accession sra_accession        date         region  \\\n",
       "0              OU420663    ERR5939958        2020         Europe   \n",
       "1              MZ702266           NaN  2020-12-02  North America   \n",
       "2              OU281100           NaN  2020-10-18         Europe   \n",
       "3              OU314699    ERR6069165  2021-06-02         Europe   \n",
       "4              OU329920    ERR6085250  2021-06-10         Europe   \n",
       "...                 ...           ...         ...            ...   \n",
       "99995          MW989219           NaN  2021-04-05  North America   \n",
       "99996          MW704675           NaN  2021-02-09  North America   \n",
       "99997          MW689132           NaN  2020-11-06  North America   \n",
       "99998          OD978562    ERR5303908  2021-01-17         Europe   \n",
       "99999          OB997017    ERR4877748  2020-11-08         Europe   \n",
       "\n",
       "              country    division             location  ... divergence  \\\n",
       "0      United Kingdom     England                  NaN  ...       40.0   \n",
       "1                 USA    New York        New York City  ...       21.0   \n",
       "2         Switzerland  Basel-Land                  NaN  ...       19.0   \n",
       "3      United Kingdom    Scotland                  NaN  ...       33.0   \n",
       "4      United Kingdom    Scotland                  NaN  ...       37.0   \n",
       "...               ...         ...                  ...  ...        ...   \n",
       "99995             USA        Ohio                  NaN  ...       35.0   \n",
       "99996             USA      Kansas                  NaN  ...       21.0   \n",
       "99997             USA  Washington  Grays Harbor County  ...       14.0   \n",
       "99998  United Kingdom     England                  NaN  ...       32.0   \n",
       "99999  United Kingdom     England                  NaN  ...       16.0   \n",
       "\n",
       "      nonACGTN rare_mutations snp_clusters  QC_missing_data QC_mixed_sites  \\\n",
       "0          0.0           15.0          0.0             good           good   \n",
       "1          0.0            7.0          0.0         mediocre           good   \n",
       "2          0.0            6.0          0.0             good           good   \n",
       "3          0.0            7.0          0.0         mediocre           good   \n",
       "4          0.0            5.0          0.0             good           good   \n",
       "...        ...            ...          ...              ...            ...   \n",
       "99995      0.0            9.0          0.0             good           good   \n",
       "99996      0.0            6.0          0.0             good           good   \n",
       "99997      0.0            0.0          0.0             good           good   \n",
       "99998      0.0            7.0          0.0             good           good   \n",
       "99999      0.0            3.0          0.0             good           good   \n",
       "\n",
       "      QC_rare_mutations QC_snp_clusters clock_deviation missing_data_percent  \n",
       "0                  good            good               ?             0.351135  \n",
       "1                  good            good             0.0             4.417617  \n",
       "2                  good            good             0.0             0.297629  \n",
       "3                  good            good            -5.0             3.885898  \n",
       "4                  good            good             1.0             0.404642  \n",
       "...                 ...             ...             ...                  ...  \n",
       "99995              good            good             0.0             0.000000  \n",
       "99996              good            good            -2.0             3.394308  \n",
       "99997              good            good            -2.0             0.006688  \n",
       "99998              good            good             0.0             1.595158  \n",
       "99999              good            good            -1.0             2.324182  \n",
       "\n",
       "[100000 rows x 40 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "metadata_df = pd.read_csv(subset_metadata_path, sep='\\t')\n",
    "metadata_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "7fc53361-b9da-4763-89ed-1ee8332f24d1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['strain', 'virus', 'gisaid_epi_isl', 'genbank_accession',\n",
       "       'sra_accession', 'date', 'region', 'country', 'division', 'location',\n",
       "       'region_exposure', 'country_exposure', 'division_exposure', 'segment',\n",
       "       'length', 'host', 'age', 'sex', 'Nextstrain_clade', 'pango_lineage',\n",
       "       'GISAID_clade', 'originating_lab', 'submitting_lab', 'authors', 'url',\n",
       "       'title', 'paper_url', 'date_submitted', 'sampling_strategy',\n",
       "       'missing_data', 'divergence', 'nonACGTN', 'rare_mutations',\n",
       "       'snp_clusters', 'QC_missing_data', 'QC_mixed_sites',\n",
       "       'QC_rare_mutations', 'QC_snp_clusters', 'clock_deviation',\n",
       "       'missing_data_percent'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata_df.columns"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "580ea56f-1a60-4382-8eb7-e7db9b1650eb",
   "metadata": {},
   "source": [
    "## 3.3. Join to pangolin results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "94a19e11-f3ad-4206-8845-e1299a018115",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>strain</th>\n",
       "      <th>virus</th>\n",
       "      <th>gisaid_epi_isl</th>\n",
       "      <th>genbank_accession</th>\n",
       "      <th>sra_accession</th>\n",
       "      <th>date</th>\n",
       "      <th>region</th>\n",
       "      <th>country</th>\n",
       "      <th>division</th>\n",
       "      <th>location</th>\n",
       "      <th>...</th>\n",
       "      <th>ambiguity_score</th>\n",
       "      <th>scorpio_call</th>\n",
       "      <th>scorpio_support</th>\n",
       "      <th>scorpio_conflict</th>\n",
       "      <th>version</th>\n",
       "      <th>pangolin_version</th>\n",
       "      <th>pangoLEARN_version</th>\n",
       "      <th>pango_version</th>\n",
       "      <th>status</th>\n",
       "      <th>note</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>OU420663</td>\n",
       "      <td>ncov</td>\n",
       "      <td>?</td>\n",
       "      <td>OU420663</td>\n",
       "      <td>ERR5939958</td>\n",
       "      <td>2020</td>\n",
       "      <td>Europe</td>\n",
       "      <td>United Kingdom</td>\n",
       "      <td>England</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>Alpha (B.1.1.7-like)</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>PLEARN-v1.2.127</td>\n",
       "      <td>3.1.20</td>\n",
       "      <td>2022-02-28</td>\n",
       "      <td>v1.2.127</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td>scorpio call: Alt alleles 23; Ref alleles 0; A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>USA/NY-CUIMC-NP-3606/2020</td>\n",
       "      <td>ncov</td>\n",
       "      <td>?</td>\n",
       "      <td>MZ702266</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2020-12-02</td>\n",
       "      <td>North America</td>\n",
       "      <td>USA</td>\n",
       "      <td>New York</td>\n",
       "      <td>New York City</td>\n",
       "      <td>...</td>\n",
       "      <td>0.951273</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PLEARN-v1.2.127</td>\n",
       "      <td>3.1.20</td>\n",
       "      <td>2022-02-28</td>\n",
       "      <td>v1.2.127</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Switzerland/BL-UHB-42491849/2020</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_930932</td>\n",
       "      <td>OU281100</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2020-10-18</td>\n",
       "      <td>Europe</td>\n",
       "      <td>Switzerland</td>\n",
       "      <td>Basel-Land</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PANGO-v1.2.127</td>\n",
       "      <td>3.1.20</td>\n",
       "      <td>2022-02-28</td>\n",
       "      <td>v1.2.127</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td>Assigned from designation hash.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Scotland/QEUH-1618558/2021</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_2487678</td>\n",
       "      <td>OU314699</td>\n",
       "      <td>ERR6069165</td>\n",
       "      <td>2021-06-02</td>\n",
       "      <td>Europe</td>\n",
       "      <td>United Kingdom</td>\n",
       "      <td>Scotland</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.970092</td>\n",
       "      <td>Alpha (B.1.1.7-like)</td>\n",
       "      <td>0.8696</td>\n",
       "      <td>0.0435</td>\n",
       "      <td>PLEARN-v1.2.127</td>\n",
       "      <td>3.1.20</td>\n",
       "      <td>2022-02-28</td>\n",
       "      <td>v1.2.127</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td>scorpio call: Alt alleles 20; Ref alleles 1; A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Scotland/QEUH-16829C6/2021</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_2554507</td>\n",
       "      <td>OU329920</td>\n",
       "      <td>ERR6085250</td>\n",
       "      <td>2021-06-10</td>\n",
       "      <td>Europe</td>\n",
       "      <td>United Kingdom</td>\n",
       "      <td>Scotland</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Delta (AY.4-like)</td>\n",
       "      <td>0.9688</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>PANGO-v1.2.127</td>\n",
       "      <td>3.1.20</td>\n",
       "      <td>2022-02-28</td>\n",
       "      <td>v1.2.127</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td>scorpio call: Alt alleles 31; Ref alleles 0; A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99995</th>\n",
       "      <td>USA/OH-CDC-QDX23654955/2021</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_1669162</td>\n",
       "      <td>MW989219</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2021-04-05</td>\n",
       "      <td>North America</td>\n",
       "      <td>USA</td>\n",
       "      <td>Ohio</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>Alpha (B.1.1.7-like)</td>\n",
       "      <td>0.9565</td>\n",
       "      <td>0.0435</td>\n",
       "      <td>PLEARN-v1.2.127</td>\n",
       "      <td>3.1.20</td>\n",
       "      <td>2022-02-28</td>\n",
       "      <td>v1.2.127</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td>scorpio call: Alt alleles 22; Ref alleles 1; A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99996</th>\n",
       "      <td>USA/KS-CDC-LC0014845/2021</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_1255089</td>\n",
       "      <td>MW704675</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2021-02-09</td>\n",
       "      <td>North America</td>\n",
       "      <td>USA</td>\n",
       "      <td>Kansas</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.976650</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PLEARN-v1.2.127</td>\n",
       "      <td>3.1.20</td>\n",
       "      <td>2022-02-28</td>\n",
       "      <td>v1.2.127</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99997</th>\n",
       "      <td>USA/WA-S5065/2020</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_1114222</td>\n",
       "      <td>MW689132</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2020-11-06</td>\n",
       "      <td>North America</td>\n",
       "      <td>USA</td>\n",
       "      <td>Washington</td>\n",
       "      <td>Grays Harbor County</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PLEARN-v1.2.127</td>\n",
       "      <td>3.1.20</td>\n",
       "      <td>2022-02-28</td>\n",
       "      <td>v1.2.127</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99998</th>\n",
       "      <td>England/MILK-10AE881/2021</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_993691</td>\n",
       "      <td>OD978562</td>\n",
       "      <td>ERR5303908</td>\n",
       "      <td>2021-01-17</td>\n",
       "      <td>Europe</td>\n",
       "      <td>United Kingdom</td>\n",
       "      <td>England</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Alpha (B.1.1.7-like)</td>\n",
       "      <td>0.9565</td>\n",
       "      <td>0.0435</td>\n",
       "      <td>PANGO-v1.2.127</td>\n",
       "      <td>3.1.20</td>\n",
       "      <td>2022-02-28</td>\n",
       "      <td>v1.2.127</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td>scorpio call: Alt alleles 22; Ref alleles 1; A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99999</th>\n",
       "      <td>England/ALDP-B5B038/2020</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_655604</td>\n",
       "      <td>OB997017</td>\n",
       "      <td>ERR4877748</td>\n",
       "      <td>2020-11-08</td>\n",
       "      <td>Europe</td>\n",
       "      <td>United Kingdom</td>\n",
       "      <td>England</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PANGO-v1.2.127</td>\n",
       "      <td>3.1.20</td>\n",
       "      <td>2022-02-28</td>\n",
       "      <td>v1.2.127</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td>Assigned from designation hash.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100000 rows × 53 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 strain virus   gisaid_epi_isl  \\\n",
       "0                              OU420663  ncov                ?   \n",
       "1             USA/NY-CUIMC-NP-3606/2020  ncov                ?   \n",
       "2      Switzerland/BL-UHB-42491849/2020  ncov   EPI_ISL_930932   \n",
       "3            Scotland/QEUH-1618558/2021  ncov  EPI_ISL_2487678   \n",
       "4            Scotland/QEUH-16829C6/2021  ncov  EPI_ISL_2554507   \n",
       "...                                 ...   ...              ...   \n",
       "99995       USA/OH-CDC-QDX23654955/2021  ncov  EPI_ISL_1669162   \n",
       "99996         USA/KS-CDC-LC0014845/2021  ncov  EPI_ISL_1255089   \n",
       "99997                 USA/WA-S5065/2020  ncov  EPI_ISL_1114222   \n",
       "99998         England/MILK-10AE881/2021  ncov   EPI_ISL_993691   \n",
       "99999          England/ALDP-B5B038/2020  ncov   EPI_ISL_655604   \n",
       "\n",
       "      genbank_accession sra_accession        date         region  \\\n",
       "0              OU420663    ERR5939958        2020         Europe   \n",
       "1              MZ702266           NaN  2020-12-02  North America   \n",
       "2              OU281100           NaN  2020-10-18         Europe   \n",
       "3              OU314699    ERR6069165  2021-06-02         Europe   \n",
       "4              OU329920    ERR6085250  2021-06-10         Europe   \n",
       "...                 ...           ...         ...            ...   \n",
       "99995          MW989219           NaN  2021-04-05  North America   \n",
       "99996          MW704675           NaN  2021-02-09  North America   \n",
       "99997          MW689132           NaN  2020-11-06  North America   \n",
       "99998          OD978562    ERR5303908  2021-01-17         Europe   \n",
       "99999          OB997017    ERR4877748  2020-11-08         Europe   \n",
       "\n",
       "              country    division             location  ... ambiguity_score  \\\n",
       "0      United Kingdom     England                  NaN  ...        1.000000   \n",
       "1                 USA    New York        New York City  ...        0.951273   \n",
       "2         Switzerland  Basel-Land                  NaN  ...             NaN   \n",
       "3      United Kingdom    Scotland                  NaN  ...        0.970092   \n",
       "4      United Kingdom    Scotland                  NaN  ...             NaN   \n",
       "...               ...         ...                  ...  ...             ...   \n",
       "99995             USA        Ohio                  NaN  ...        1.000000   \n",
       "99996             USA      Kansas                  NaN  ...        0.976650   \n",
       "99997             USA  Washington  Grays Harbor County  ...        1.000000   \n",
       "99998  United Kingdom     England                  NaN  ...             NaN   \n",
       "99999  United Kingdom     England                  NaN  ...             NaN   \n",
       "\n",
       "               scorpio_call scorpio_support scorpio_conflict          version  \\\n",
       "0      Alpha (B.1.1.7-like)          1.0000           0.0000  PLEARN-v1.2.127   \n",
       "1                       NaN             NaN              NaN  PLEARN-v1.2.127   \n",
       "2                       NaN             NaN              NaN   PANGO-v1.2.127   \n",
       "3      Alpha (B.1.1.7-like)          0.8696           0.0435  PLEARN-v1.2.127   \n",
       "4         Delta (AY.4-like)          0.9688           0.0000   PANGO-v1.2.127   \n",
       "...                     ...             ...              ...              ...   \n",
       "99995  Alpha (B.1.1.7-like)          0.9565           0.0435  PLEARN-v1.2.127   \n",
       "99996                   NaN             NaN              NaN  PLEARN-v1.2.127   \n",
       "99997                   NaN             NaN              NaN  PLEARN-v1.2.127   \n",
       "99998  Alpha (B.1.1.7-like)          0.9565           0.0435   PANGO-v1.2.127   \n",
       "99999                   NaN             NaN              NaN   PANGO-v1.2.127   \n",
       "\n",
       "      pangolin_version pangoLEARN_version pango_version     status  \\\n",
       "0               3.1.20         2022-02-28      v1.2.127  passed_qc   \n",
       "1               3.1.20         2022-02-28      v1.2.127  passed_qc   \n",
       "2               3.1.20         2022-02-28      v1.2.127  passed_qc   \n",
       "3               3.1.20         2022-02-28      v1.2.127  passed_qc   \n",
       "4               3.1.20         2022-02-28      v1.2.127  passed_qc   \n",
       "...                ...                ...           ...        ...   \n",
       "99995           3.1.20         2022-02-28      v1.2.127  passed_qc   \n",
       "99996           3.1.20         2022-02-28      v1.2.127  passed_qc   \n",
       "99997           3.1.20         2022-02-28      v1.2.127  passed_qc   \n",
       "99998           3.1.20         2022-02-28      v1.2.127  passed_qc   \n",
       "99999           3.1.20         2022-02-28      v1.2.127  passed_qc   \n",
       "\n",
       "                                                    note  \n",
       "0      scorpio call: Alt alleles 23; Ref alleles 0; A...  \n",
       "1                                                    NaN  \n",
       "2                        Assigned from designation hash.  \n",
       "3      scorpio call: Alt alleles 20; Ref alleles 1; A...  \n",
       "4      scorpio call: Alt alleles 31; Ref alleles 0; A...  \n",
       "...                                                  ...  \n",
       "99995  scorpio call: Alt alleles 22; Ref alleles 1; A...  \n",
       "99996                                                NaN  \n",
       "99997                                                NaN  \n",
       "99998  scorpio call: Alt alleles 22; Ref alleles 1; A...  \n",
       "99999                    Assigned from designation hash.  \n",
       "\n",
       "[100000 rows x 53 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata_pangolin_df = metadata_df.merge(p_df, left_on='strain', right_on='taxon')\n",
    "metadata_pangolin_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "58d82e66-28b1-4882-8b28-b422eacf3041",
   "metadata": {},
   "source": [
    "# 4. Save joined metadata to file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "9761daae-e204-42d7-84fc-9e49b775fde7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saved metadata and pangolin to input/metadata-subsample-pangolin.tsv\n"
     ]
    }
   ],
   "source": [
    "metadata_pangolin_df.to_csv(output_metadata_pangolin_path, sep='\\t', index=False)\n",
    "\n",
    "print(f'Saved metadata and pangolin to {output_metadata_pangolin_path}')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
