{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "cadee23e-1b03-4332-8d60-9c1239f30f8e",
   "metadata": {},
   "source": [
    "# 1. Parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "23e047ea-2e3f-4a42-867b-7112a179f261",
   "metadata": {},
   "outputs": [],
   "source": [
    "sequences_split_dir = 'input/split'\n",
    "sequences_subset_all = 'input/all.fasta.gz'\n",
    "pangolin_out_dir = 'input/pangolin'\n",
    "subset_metadata_path = 'input/metadata-subsample.tsv'\n",
    "output_metadata_pangolin_path = 'input/metadata-subsample-pangolin.tsv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ad50bedb-8447-4c3c-b407-f1812734aab7",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "\n",
    "sequences_split_dir = Path(sequences_split_dir)\n",
    "sequences_subset_all = Path(sequences_subset_all)\n",
    "pangolin_out_dir = Path(pangolin_out_dir)\n",
    "subset_metadata_path = Path(subset_metadata_path)\n",
    "output_metadata_pangolin_path = Path(output_metadata_pangolin_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "73ccbafc-81c6-41e9-962c-014ee21fd6bd",
   "metadata": {},
   "source": [
    "# 2. Run pangolin"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "e88ec771-2ba6-48a1-8d72-c1ba8d667dc3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pangolin: 3.1.11\n",
      "pangolearn: 2021-08-24\n",
      "constellations: v0.0.15\n",
      "scorpio: 0.3.12\n",
      "pango-designation: 1.2.76\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!conda run --name pangolin pangolin --all-versions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "fa9413c2-a261-4ade-9789-551f33132ba4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "loading model 09/07/2021, 21:27:17\n",
      "processing block of 1000 sequences 09/07/2021, 21:27:28\n",
      "processing block of 1000 sequences 09/07/2021, 21:27:57\n",
      "processing block of 1000 sequences 09/07/2021, 21:28:21\n",
      "processing block of 1000 sequences 09/07/2021, 21:28:45\n",
      "processing block of 1000 sequences 09/07/2021, 21:29:08\n",
      "processing block of 1000 sequences 09/07/2021, 21:29:32\n",
      "processing block of 1000 sequences 09/07/2021, 21:29:55\n",
      "processing block of 139 sequences 09/07/2021, 21:30:10\n",
      "complete 09/07/2021, 21:30:13\n",
      "\u001b[32mOutput file written to: \u001b[0m/home/CSCScience.ca/apetkau/workspace/genomics-data-index-evaluation/evaluations/sars-cov-2/input/pangolin/lineage_report.csv\n",
      "\u001b[32mAll dependencies satisfied.\u001b[0m\n",
      "\u001b[32mThe query file is:\u001b[0m/home/CSCScience.ca/apetkau/workspace/genomics-data-index-evaluation/evaluations/sars-cov-2/input/all.fasta.gz\n",
      "\u001b[32m** Sequence QC **\u001b[0m\n",
      "Sequence name                 \t                   Reason\tValue     \n",
      "\n",
      "\u001b[32m\n",
      "Number of sequences detected: \u001b[0m10000\n",
      "\u001b[32mTotal passing QC: \u001b[0m10000\n",
      "\u001b[32m\n",
      "Data files found:\u001b[0m\n",
      "Trained model:\t/home/CSCScience.ca/apetkau/miniconda3/envs/pangolin/lib/python3.8/site-packages/pangoLEARN/data/decisionTree_v1.joblib\n",
      "Header file:\t/home/CSCScience.ca/apetkau/miniconda3/envs/pangolin/lib/python3.8/site-packages/pangoLEARN/data/decisionTreeHeaders_v1.joblib\n",
      "Designated hash:\t/home/CSCScience.ca/apetkau/miniconda3/envs/pangolin/lib/python3.8/site-packages/pangoLEARN/data/lineages.hash.csv\n",
      "\n",
      "Job stats:\n",
      "job                     count    min threads    max threads\n",
      "--------------------  -------  -------------  -------------\n",
      "add_failed_seqs             1              1              1\n",
      "align_to_reference          1              1              1\n",
      "all                         1              1              1\n",
      "generate_report             1              1              1\n",
      "get_constellations          1              1              1\n",
      "hash_sequence_assign        1              1              1\n",
      "pangolearn                  1              1              1\n",
      "scorpio                     1             52             52\n",
      "total                       8              1             52\n",
      "\n",
      "/home/CSCScience.ca/apetkau/miniconda3/envs/pangolin/lib/python3.8/site-packages/sklearn/base.py:329: UserWarning: Trying to unpickle estimator DecisionTreeClassifier from version 0.24.2 when using version 0.23.1. This might lead to breaking code or invalid results. Use at your own risk.\n",
      "  warnings.warn(\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!conda run --name pangolin pangolin -t 52 -o {pangolin_out_dir} {sequences_subset_all}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4897c5b3-bd6c-445f-a156-153ad3b97120",
   "metadata": {},
   "source": [
    "# 3. Load up and join pangolin to metadata\n",
    "\n",
    "## 3.1. Load pangolin results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c2e59423-ee34-4303-bc55-f44a9bca8a63",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>taxon</th>\n",
       "      <th>lineage</th>\n",
       "      <th>conflict</th>\n",
       "      <th>ambiguity_score</th>\n",
       "      <th>scorpio_call</th>\n",
       "      <th>scorpio_support</th>\n",
       "      <th>scorpio_conflict</th>\n",
       "      <th>version</th>\n",
       "      <th>pangolin_version</th>\n",
       "      <th>pangoLEARN_version</th>\n",
       "      <th>pango_version</th>\n",
       "      <th>status</th>\n",
       "      <th>note</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>LR862463</td>\n",
       "      <td>C.35</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PANGO-v1.2.66</td>\n",
       "      <td>3.1.11</td>\n",
       "      <td>2021-08-24</td>\n",
       "      <td>v1.2.66</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td>Assigned from designation hash.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>LR882087</td>\n",
       "      <td>B.1.177.23</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PANGO-v1.2.66</td>\n",
       "      <td>3.1.11</td>\n",
       "      <td>2021-08-24</td>\n",
       "      <td>v1.2.66</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td>Assigned from designation hash.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>LR882100</td>\n",
       "      <td>B.1.1.70</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PANGO-v1.2.66</td>\n",
       "      <td>3.1.11</td>\n",
       "      <td>2021-08-24</td>\n",
       "      <td>v1.2.66</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td>Assigned from designation hash.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>LR882436</td>\n",
       "      <td>B.1.1.39</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PANGO-v1.2.66</td>\n",
       "      <td>3.1.11</td>\n",
       "      <td>2021-08-24</td>\n",
       "      <td>v1.2.66</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td>Assigned from designation hash.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LR882838</td>\n",
       "      <td>B.1.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PANGO-v1.2.66</td>\n",
       "      <td>3.1.11</td>\n",
       "      <td>2021-08-24</td>\n",
       "      <td>v1.2.66</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td>Assigned from designation hash.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9995</th>\n",
       "      <td>OU149351</td>\n",
       "      <td>B.1.1.7</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.999851</td>\n",
       "      <td>Alpha (B.1.1.7-like)</td>\n",
       "      <td>0.9565</td>\n",
       "      <td>0.0435</td>\n",
       "      <td>PLEARN-v1.2.66</td>\n",
       "      <td>3.1.11</td>\n",
       "      <td>2021-08-24</td>\n",
       "      <td>v1.2.66</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td>scorpio call: Alt alleles 22; Ref alleles 1; A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9996</th>\n",
       "      <td>OU149407</td>\n",
       "      <td>B.1.1.7</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.992859</td>\n",
       "      <td>Alpha (B.1.1.7-like)</td>\n",
       "      <td>0.9565</td>\n",
       "      <td>0.0435</td>\n",
       "      <td>PLEARN-v1.2.66</td>\n",
       "      <td>3.1.11</td>\n",
       "      <td>2021-08-24</td>\n",
       "      <td>v1.2.66</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td>scorpio call: Alt alleles 22; Ref alleles 1; A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9997</th>\n",
       "      <td>OU149436</td>\n",
       "      <td>B.1.1.7</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.999666</td>\n",
       "      <td>Alpha (B.1.1.7-like)</td>\n",
       "      <td>0.8696</td>\n",
       "      <td>0.0435</td>\n",
       "      <td>PLEARN-v1.2.66</td>\n",
       "      <td>3.1.11</td>\n",
       "      <td>2021-08-24</td>\n",
       "      <td>v1.2.66</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td>scorpio call: Alt alleles 20; Ref alleles 1; A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9998</th>\n",
       "      <td>OU149604</td>\n",
       "      <td>B.1.1.7</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>Alpha (B.1.1.7-like)</td>\n",
       "      <td>0.9565</td>\n",
       "      <td>0.0435</td>\n",
       "      <td>PLEARN-v1.2.66</td>\n",
       "      <td>3.1.11</td>\n",
       "      <td>2021-08-24</td>\n",
       "      <td>v1.2.66</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td>scorpio call: Alt alleles 22; Ref alleles 1; A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9999</th>\n",
       "      <td>OU149613</td>\n",
       "      <td>B.1.1.7</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>Alpha (B.1.1.7-like)</td>\n",
       "      <td>0.9565</td>\n",
       "      <td>0.0435</td>\n",
       "      <td>PLEARN-v1.2.66</td>\n",
       "      <td>3.1.11</td>\n",
       "      <td>2021-08-24</td>\n",
       "      <td>v1.2.66</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td>scorpio call: Alt alleles 22; Ref alleles 1; A...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10000 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         taxon     lineage  conflict  ambiguity_score          scorpio_call  \\\n",
       "0     LR862463        C.35       NaN              NaN                   NaN   \n",
       "1     LR882087  B.1.177.23       NaN              NaN                   NaN   \n",
       "2     LR882100    B.1.1.70       NaN              NaN                   NaN   \n",
       "3     LR882436    B.1.1.39       NaN              NaN                   NaN   \n",
       "4     LR882838       B.1.1       NaN              NaN                   NaN   \n",
       "...        ...         ...       ...              ...                   ...   \n",
       "9995  OU149351     B.1.1.7       0.0         0.999851  Alpha (B.1.1.7-like)   \n",
       "9996  OU149407     B.1.1.7       0.0         0.992859  Alpha (B.1.1.7-like)   \n",
       "9997  OU149436     B.1.1.7       0.0         0.999666  Alpha (B.1.1.7-like)   \n",
       "9998  OU149604     B.1.1.7       0.0         1.000000  Alpha (B.1.1.7-like)   \n",
       "9999  OU149613     B.1.1.7       0.0         1.000000  Alpha (B.1.1.7-like)   \n",
       "\n",
       "      scorpio_support  scorpio_conflict         version pangolin_version  \\\n",
       "0                 NaN               NaN   PANGO-v1.2.66           3.1.11   \n",
       "1                 NaN               NaN   PANGO-v1.2.66           3.1.11   \n",
       "2                 NaN               NaN   PANGO-v1.2.66           3.1.11   \n",
       "3                 NaN               NaN   PANGO-v1.2.66           3.1.11   \n",
       "4                 NaN               NaN   PANGO-v1.2.66           3.1.11   \n",
       "...               ...               ...             ...              ...   \n",
       "9995           0.9565            0.0435  PLEARN-v1.2.66           3.1.11   \n",
       "9996           0.9565            0.0435  PLEARN-v1.2.66           3.1.11   \n",
       "9997           0.8696            0.0435  PLEARN-v1.2.66           3.1.11   \n",
       "9998           0.9565            0.0435  PLEARN-v1.2.66           3.1.11   \n",
       "9999           0.9565            0.0435  PLEARN-v1.2.66           3.1.11   \n",
       "\n",
       "     pangoLEARN_version pango_version     status  \\\n",
       "0            2021-08-24       v1.2.66  passed_qc   \n",
       "1            2021-08-24       v1.2.66  passed_qc   \n",
       "2            2021-08-24       v1.2.66  passed_qc   \n",
       "3            2021-08-24       v1.2.66  passed_qc   \n",
       "4            2021-08-24       v1.2.66  passed_qc   \n",
       "...                 ...           ...        ...   \n",
       "9995         2021-08-24       v1.2.66  passed_qc   \n",
       "9996         2021-08-24       v1.2.66  passed_qc   \n",
       "9997         2021-08-24       v1.2.66  passed_qc   \n",
       "9998         2021-08-24       v1.2.66  passed_qc   \n",
       "9999         2021-08-24       v1.2.66  passed_qc   \n",
       "\n",
       "                                                   note  \n",
       "0                       Assigned from designation hash.  \n",
       "1                       Assigned from designation hash.  \n",
       "2                       Assigned from designation hash.  \n",
       "3                       Assigned from designation hash.  \n",
       "4                       Assigned from designation hash.  \n",
       "...                                                 ...  \n",
       "9995  scorpio call: Alt alleles 22; Ref alleles 1; A...  \n",
       "9996  scorpio call: Alt alleles 22; Ref alleles 1; A...  \n",
       "9997  scorpio call: Alt alleles 20; Ref alleles 1; A...  \n",
       "9998  scorpio call: Alt alleles 22; Ref alleles 1; A...  \n",
       "9999  scorpio call: Alt alleles 22; Ref alleles 1; A...  \n",
       "\n",
       "[10000 rows x 13 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "pangolin_report = pangolin_out_dir / 'lineage_report.csv'\n",
    "\n",
    "p_df = pd.read_csv(pangolin_report)\n",
    "p_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d199cae3-aa10-4d48-948e-209a2b38f726",
   "metadata": {},
   "source": [
    "## 3.2. Load other metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "bacd0e35-2d5d-48d1-90ed-329a05700367",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>strain</th>\n",
       "      <th>virus</th>\n",
       "      <th>gisaid_epi_isl</th>\n",
       "      <th>genbank_accession</th>\n",
       "      <th>sra_accession</th>\n",
       "      <th>date</th>\n",
       "      <th>region</th>\n",
       "      <th>country</th>\n",
       "      <th>division</th>\n",
       "      <th>location</th>\n",
       "      <th>...</th>\n",
       "      <th>divergence</th>\n",
       "      <th>nonACGTN</th>\n",
       "      <th>rare_mutations</th>\n",
       "      <th>snp_clusters</th>\n",
       "      <th>QC_missing_data</th>\n",
       "      <th>QC_mixed_sites</th>\n",
       "      <th>QC_rare_mutations</th>\n",
       "      <th>QC_snp_clusters</th>\n",
       "      <th>clock_deviation</th>\n",
       "      <th>missing_data_percent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>England/QEUH-148725F/2021</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_1487327</td>\n",
       "      <td>OU224813</td>\n",
       "      <td>ERR5679447</td>\n",
       "      <td>2021-03-28</td>\n",
       "      <td>Europe</td>\n",
       "      <td>United Kingdom</td>\n",
       "      <td>England</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>34.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.404642</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>England/MILK-1535AC0/2021</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_1985652</td>\n",
       "      <td>OU266415</td>\n",
       "      <td>ERR5905912</td>\n",
       "      <td>2021-04-24</td>\n",
       "      <td>Europe</td>\n",
       "      <td>United Kingdom</td>\n",
       "      <td>England</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>31.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.404642</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>England/QEUH-789A28/2020</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_533176</td>\n",
       "      <td>OA970628</td>\n",
       "      <td>ERR4584389</td>\n",
       "      <td>2020-07-15</td>\n",
       "      <td>Europe</td>\n",
       "      <td>United Kingdom</td>\n",
       "      <td>England</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>14.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.404642</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>USA/CA-CDC-ASC210034758/2021</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_2042505</td>\n",
       "      <td>MZ184110</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2021-04-26</td>\n",
       "      <td>North America</td>\n",
       "      <td>USA</td>\n",
       "      <td>California</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>37.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.016721</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Switzerland/ZH-ETHZ-500086/2020</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_1119047</td>\n",
       "      <td>OU011535</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2020-11-09</td>\n",
       "      <td>Europe</td>\n",
       "      <td>Switzerland</td>\n",
       "      <td>Zürich</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>32.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9995</th>\n",
       "      <td>OU201734</td>\n",
       "      <td>ncov</td>\n",
       "      <td>?</td>\n",
       "      <td>OU201734</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2021-05-08</td>\n",
       "      <td>Europe</td>\n",
       "      <td>Germany</td>\n",
       "      <td>North Rhine Westphalia</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>32.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>good</td>\n",
       "      <td>mediocre</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>-5.0</td>\n",
       "      <td>0.678862</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9996</th>\n",
       "      <td>England/ALDP-18C9AA0/2021</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_2973222</td>\n",
       "      <td>OU436787</td>\n",
       "      <td>ERR6327095</td>\n",
       "      <td>2021-07-07</td>\n",
       "      <td>Europe</td>\n",
       "      <td>United Kingdom</td>\n",
       "      <td>England</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>39.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.404642</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9997</th>\n",
       "      <td>England/RAND-12BEEC6/2021</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_1070245</td>\n",
       "      <td>OD996254</td>\n",
       "      <td>ERR5380596</td>\n",
       "      <td>2021-02-10</td>\n",
       "      <td>Europe</td>\n",
       "      <td>United Kingdom</td>\n",
       "      <td>England</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>30.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>-2.0</td>\n",
       "      <td>0.404642</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9998</th>\n",
       "      <td>USA/MA-CDC-ASC210073741/2021</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_2041225</td>\n",
       "      <td>MZ187570</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2021-04-17</td>\n",
       "      <td>North America</td>\n",
       "      <td>USA</td>\n",
       "      <td>Massachusetts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>37.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9999</th>\n",
       "      <td>England/ALDP-1A019CB/2021</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_3168561</td>\n",
       "      <td>OU477474</td>\n",
       "      <td>ERR6403991</td>\n",
       "      <td>2021-07-23</td>\n",
       "      <td>Europe</td>\n",
       "      <td>United Kingdom</td>\n",
       "      <td>England</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>38.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>good</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.651908</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10000 rows × 40 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                               strain virus   gisaid_epi_isl  \\\n",
       "0           England/QEUH-148725F/2021  ncov  EPI_ISL_1487327   \n",
       "1           England/MILK-1535AC0/2021  ncov  EPI_ISL_1985652   \n",
       "2            England/QEUH-789A28/2020  ncov   EPI_ISL_533176   \n",
       "3        USA/CA-CDC-ASC210034758/2021  ncov  EPI_ISL_2042505   \n",
       "4     Switzerland/ZH-ETHZ-500086/2020  ncov  EPI_ISL_1119047   \n",
       "...                               ...   ...              ...   \n",
       "9995                         OU201734  ncov                ?   \n",
       "9996        England/ALDP-18C9AA0/2021  ncov  EPI_ISL_2973222   \n",
       "9997        England/RAND-12BEEC6/2021  ncov  EPI_ISL_1070245   \n",
       "9998     USA/MA-CDC-ASC210073741/2021  ncov  EPI_ISL_2041225   \n",
       "9999        England/ALDP-1A019CB/2021  ncov  EPI_ISL_3168561   \n",
       "\n",
       "     genbank_accession sra_accession        date         region  \\\n",
       "0             OU224813    ERR5679447  2021-03-28         Europe   \n",
       "1             OU266415    ERR5905912  2021-04-24         Europe   \n",
       "2             OA970628    ERR4584389  2020-07-15         Europe   \n",
       "3             MZ184110           NaN  2021-04-26  North America   \n",
       "4             OU011535           NaN  2020-11-09         Europe   \n",
       "...                ...           ...         ...            ...   \n",
       "9995          OU201734           NaN  2021-05-08         Europe   \n",
       "9996          OU436787    ERR6327095  2021-07-07         Europe   \n",
       "9997          OD996254    ERR5380596  2021-02-10         Europe   \n",
       "9998          MZ187570           NaN  2021-04-17  North America   \n",
       "9999          OU477474    ERR6403991  2021-07-23         Europe   \n",
       "\n",
       "             country                division location  ... divergence  \\\n",
       "0     United Kingdom                 England      NaN  ...       34.0   \n",
       "1     United Kingdom                 England      NaN  ...       31.0   \n",
       "2     United Kingdom                 England      NaN  ...       14.0   \n",
       "3                USA              California      NaN  ...       37.0   \n",
       "4        Switzerland                  Zürich      NaN  ...       32.0   \n",
       "...              ...                     ...      ...  ...        ...   \n",
       "9995         Germany  North Rhine Westphalia      NaN  ...       32.0   \n",
       "9996  United Kingdom                 England      NaN  ...       39.0   \n",
       "9997  United Kingdom                 England      NaN  ...       30.0   \n",
       "9998             USA           Massachusetts      NaN  ...       37.0   \n",
       "9999  United Kingdom                 England      NaN  ...       38.0   \n",
       "\n",
       "     nonACGTN rare_mutations snp_clusters  QC_missing_data QC_mixed_sites  \\\n",
       "0         0.0            9.0          0.0             good           good   \n",
       "1         0.0           10.0          0.0             good           good   \n",
       "2         0.0            7.0          0.0             good           good   \n",
       "3         0.0            7.0          0.0             good           good   \n",
       "4         0.0            7.0          0.0             good           good   \n",
       "...       ...            ...          ...              ...            ...   \n",
       "9995      3.0            8.0          0.0             good       mediocre   \n",
       "9996      0.0           10.0          0.0             good           good   \n",
       "9997      0.0            5.0          0.0             good           good   \n",
       "9998      0.0            8.0          0.0             good           good   \n",
       "9999      2.0            7.0          0.0             good           good   \n",
       "\n",
       "     QC_rare_mutations QC_snp_clusters clock_deviation missing_data_percent  \n",
       "0                 good            good            -1.0             0.404642  \n",
       "1                 good            good             1.0             0.404642  \n",
       "2                 good            good             1.0             0.404642  \n",
       "3                 good            good             0.0             0.016721  \n",
       "4                 good            good             4.0             0.000000  \n",
       "...                ...             ...             ...                  ...  \n",
       "9995              good            good            -5.0             0.678862  \n",
       "9996              good            good             2.0             0.404642  \n",
       "9997              good            good            -2.0             0.404642  \n",
       "9998              good            good             0.0             0.000000  \n",
       "9999              good            good             0.0             2.651908  \n",
       "\n",
       "[10000 rows x 40 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "metadata_df = pd.read_csv(subset_metadata_path, sep='\\t')\n",
    "metadata_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "7fc53361-b9da-4763-89ed-1ee8332f24d1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['strain', 'virus', 'gisaid_epi_isl', 'genbank_accession',\n",
       "       'sra_accession', 'date', 'region', 'country', 'division', 'location',\n",
       "       'region_exposure', 'country_exposure', 'division_exposure', 'segment',\n",
       "       'length', 'host', 'age', 'sex', 'Nextstrain_clade', 'pango_lineage',\n",
       "       'GISAID_clade', 'originating_lab', 'submitting_lab', 'authors', 'url',\n",
       "       'title', 'paper_url', 'date_submitted', 'sampling_strategy',\n",
       "       'missing_data', 'divergence', 'nonACGTN', 'rare_mutations',\n",
       "       'snp_clusters', 'QC_missing_data', 'QC_mixed_sites',\n",
       "       'QC_rare_mutations', 'QC_snp_clusters', 'clock_deviation',\n",
       "       'missing_data_percent'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata_df.columns"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "580ea56f-1a60-4382-8eb7-e7db9b1650eb",
   "metadata": {},
   "source": [
    "## 3.3. Join to pangolin results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "94a19e11-f3ad-4206-8845-e1299a018115",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>strain</th>\n",
       "      <th>virus</th>\n",
       "      <th>gisaid_epi_isl</th>\n",
       "      <th>genbank_accession</th>\n",
       "      <th>sra_accession</th>\n",
       "      <th>date</th>\n",
       "      <th>region</th>\n",
       "      <th>country</th>\n",
       "      <th>division</th>\n",
       "      <th>location</th>\n",
       "      <th>...</th>\n",
       "      <th>ambiguity_score</th>\n",
       "      <th>scorpio_call</th>\n",
       "      <th>scorpio_support</th>\n",
       "      <th>scorpio_conflict</th>\n",
       "      <th>version</th>\n",
       "      <th>pangolin_version</th>\n",
       "      <th>pangoLEARN_version</th>\n",
       "      <th>pango_version</th>\n",
       "      <th>status</th>\n",
       "      <th>note</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>England/QEUH-148725F/2021</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_1487327</td>\n",
       "      <td>OU224813</td>\n",
       "      <td>ERR5679447</td>\n",
       "      <td>2021-03-28</td>\n",
       "      <td>Europe</td>\n",
       "      <td>United Kingdom</td>\n",
       "      <td>England</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>Alpha (B.1.1.7-like)</td>\n",
       "      <td>0.9565</td>\n",
       "      <td>0.0435</td>\n",
       "      <td>PLEARN-v1.2.66</td>\n",
       "      <td>3.1.11</td>\n",
       "      <td>2021-08-24</td>\n",
       "      <td>v1.2.66</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td>scorpio call: Alt alleles 22; Ref alleles 1; A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>England/MILK-1535AC0/2021</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_1985652</td>\n",
       "      <td>OU266415</td>\n",
       "      <td>ERR5905912</td>\n",
       "      <td>2021-04-24</td>\n",
       "      <td>Europe</td>\n",
       "      <td>United Kingdom</td>\n",
       "      <td>England</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>B.1.1.318-like</td>\n",
       "      <td>0.8421</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>PLEARN-v1.2.66</td>\n",
       "      <td>3.1.11</td>\n",
       "      <td>2021-08-24</td>\n",
       "      <td>v1.2.66</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td>scorpio call: Alt alleles 16; Ref alleles 0; A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>England/QEUH-789A28/2020</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_533176</td>\n",
       "      <td>OA970628</td>\n",
       "      <td>ERR4584389</td>\n",
       "      <td>2020-07-15</td>\n",
       "      <td>Europe</td>\n",
       "      <td>United Kingdom</td>\n",
       "      <td>England</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PANGO-v1.2.66</td>\n",
       "      <td>3.1.11</td>\n",
       "      <td>2021-08-24</td>\n",
       "      <td>v1.2.66</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td>Assigned from designation hash.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>USA/CA-CDC-ASC210034758/2021</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_2042505</td>\n",
       "      <td>MZ184110</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2021-04-26</td>\n",
       "      <td>North America</td>\n",
       "      <td>USA</td>\n",
       "      <td>California</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>Gamma (P.1-like)</td>\n",
       "      <td>0.8750</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>PLEARN-v1.2.66</td>\n",
       "      <td>3.1.11</td>\n",
       "      <td>2021-08-24</td>\n",
       "      <td>v1.2.66</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td>scorpio call: Alt alleles 14; Ref alleles 0; A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Switzerland/ZH-ETHZ-500086/2020</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_1119047</td>\n",
       "      <td>OU011535</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2020-11-09</td>\n",
       "      <td>Europe</td>\n",
       "      <td>Switzerland</td>\n",
       "      <td>Zürich</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>Alpha (B.1.1.7-like)</td>\n",
       "      <td>0.9565</td>\n",
       "      <td>0.0435</td>\n",
       "      <td>PLEARN-v1.2.66</td>\n",
       "      <td>3.1.11</td>\n",
       "      <td>2021-08-24</td>\n",
       "      <td>v1.2.66</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td>scorpio call: Alt alleles 22; Ref alleles 1; A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9995</th>\n",
       "      <td>OU201734</td>\n",
       "      <td>ncov</td>\n",
       "      <td>?</td>\n",
       "      <td>OU201734</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2021-05-08</td>\n",
       "      <td>Europe</td>\n",
       "      <td>Germany</td>\n",
       "      <td>North Rhine Westphalia</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.992369</td>\n",
       "      <td>Alpha (B.1.1.7-like)</td>\n",
       "      <td>0.9565</td>\n",
       "      <td>0.0435</td>\n",
       "      <td>PLEARN-v1.2.66</td>\n",
       "      <td>3.1.11</td>\n",
       "      <td>2021-08-24</td>\n",
       "      <td>v1.2.66</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td>scorpio call: Alt alleles 22; Ref alleles 1; A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9996</th>\n",
       "      <td>England/ALDP-18C9AA0/2021</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_2973222</td>\n",
       "      <td>OU436787</td>\n",
       "      <td>ERR6327095</td>\n",
       "      <td>2021-07-07</td>\n",
       "      <td>Europe</td>\n",
       "      <td>United Kingdom</td>\n",
       "      <td>England</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Delta (B.1.617.2-like)</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>PANGO-v1.2.66</td>\n",
       "      <td>3.1.11</td>\n",
       "      <td>2021-08-24</td>\n",
       "      <td>v1.2.66</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td>scorpio call: Alt alleles 13; Ref alleles 0; A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9997</th>\n",
       "      <td>England/RAND-12BEEC6/2021</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_1070245</td>\n",
       "      <td>OD996254</td>\n",
       "      <td>ERR5380596</td>\n",
       "      <td>2021-02-10</td>\n",
       "      <td>Europe</td>\n",
       "      <td>United Kingdom</td>\n",
       "      <td>England</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Alpha (B.1.1.7-like)</td>\n",
       "      <td>0.9565</td>\n",
       "      <td>0.0435</td>\n",
       "      <td>PANGO-v1.2.66</td>\n",
       "      <td>3.1.11</td>\n",
       "      <td>2021-08-24</td>\n",
       "      <td>v1.2.66</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td>scorpio call: Alt alleles 22; Ref alleles 1; A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9998</th>\n",
       "      <td>USA/MA-CDC-ASC210073741/2021</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_2041225</td>\n",
       "      <td>MZ187570</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2021-04-17</td>\n",
       "      <td>North America</td>\n",
       "      <td>USA</td>\n",
       "      <td>Massachusetts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>Alpha (B.1.1.7-like)</td>\n",
       "      <td>0.9565</td>\n",
       "      <td>0.0435</td>\n",
       "      <td>PLEARN-v1.2.66</td>\n",
       "      <td>3.1.11</td>\n",
       "      <td>2021-08-24</td>\n",
       "      <td>v1.2.66</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td>scorpio call: Alt alleles 22; Ref alleles 1; A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9999</th>\n",
       "      <td>England/ALDP-1A019CB/2021</td>\n",
       "      <td>ncov</td>\n",
       "      <td>EPI_ISL_3168561</td>\n",
       "      <td>OU477474</td>\n",
       "      <td>ERR6403991</td>\n",
       "      <td>2021-07-23</td>\n",
       "      <td>Europe</td>\n",
       "      <td>United Kingdom</td>\n",
       "      <td>England</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.975857</td>\n",
       "      <td>Delta (B.1.617.2-like)</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>PLEARN-v1.2.66</td>\n",
       "      <td>3.1.11</td>\n",
       "      <td>2021-08-24</td>\n",
       "      <td>v1.2.66</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td>scorpio call: Alt alleles 13; Ref alleles 0; A...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10000 rows × 53 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                               strain virus   gisaid_epi_isl  \\\n",
       "0           England/QEUH-148725F/2021  ncov  EPI_ISL_1487327   \n",
       "1           England/MILK-1535AC0/2021  ncov  EPI_ISL_1985652   \n",
       "2            England/QEUH-789A28/2020  ncov   EPI_ISL_533176   \n",
       "3        USA/CA-CDC-ASC210034758/2021  ncov  EPI_ISL_2042505   \n",
       "4     Switzerland/ZH-ETHZ-500086/2020  ncov  EPI_ISL_1119047   \n",
       "...                               ...   ...              ...   \n",
       "9995                         OU201734  ncov                ?   \n",
       "9996        England/ALDP-18C9AA0/2021  ncov  EPI_ISL_2973222   \n",
       "9997        England/RAND-12BEEC6/2021  ncov  EPI_ISL_1070245   \n",
       "9998     USA/MA-CDC-ASC210073741/2021  ncov  EPI_ISL_2041225   \n",
       "9999        England/ALDP-1A019CB/2021  ncov  EPI_ISL_3168561   \n",
       "\n",
       "     genbank_accession sra_accession        date         region  \\\n",
       "0             OU224813    ERR5679447  2021-03-28         Europe   \n",
       "1             OU266415    ERR5905912  2021-04-24         Europe   \n",
       "2             OA970628    ERR4584389  2020-07-15         Europe   \n",
       "3             MZ184110           NaN  2021-04-26  North America   \n",
       "4             OU011535           NaN  2020-11-09         Europe   \n",
       "...                ...           ...         ...            ...   \n",
       "9995          OU201734           NaN  2021-05-08         Europe   \n",
       "9996          OU436787    ERR6327095  2021-07-07         Europe   \n",
       "9997          OD996254    ERR5380596  2021-02-10         Europe   \n",
       "9998          MZ187570           NaN  2021-04-17  North America   \n",
       "9999          OU477474    ERR6403991  2021-07-23         Europe   \n",
       "\n",
       "             country                division location  ... ambiguity_score  \\\n",
       "0     United Kingdom                 England      NaN  ...        1.000000   \n",
       "1     United Kingdom                 England      NaN  ...        1.000000   \n",
       "2     United Kingdom                 England      NaN  ...             NaN   \n",
       "3                USA              California      NaN  ...        1.000000   \n",
       "4        Switzerland                  Zürich      NaN  ...        1.000000   \n",
       "...              ...                     ...      ...  ...             ...   \n",
       "9995         Germany  North Rhine Westphalia      NaN  ...        0.992369   \n",
       "9996  United Kingdom                 England      NaN  ...             NaN   \n",
       "9997  United Kingdom                 England      NaN  ...             NaN   \n",
       "9998             USA           Massachusetts      NaN  ...        1.000000   \n",
       "9999  United Kingdom                 England      NaN  ...        0.975857   \n",
       "\n",
       "                scorpio_call scorpio_support scorpio_conflict         version  \\\n",
       "0       Alpha (B.1.1.7-like)          0.9565           0.0435  PLEARN-v1.2.66   \n",
       "1             B.1.1.318-like          0.8421           0.0000  PLEARN-v1.2.66   \n",
       "2                        NaN             NaN              NaN   PANGO-v1.2.66   \n",
       "3           Gamma (P.1-like)          0.8750           0.0000  PLEARN-v1.2.66   \n",
       "4       Alpha (B.1.1.7-like)          0.9565           0.0435  PLEARN-v1.2.66   \n",
       "...                      ...             ...              ...             ...   \n",
       "9995    Alpha (B.1.1.7-like)          0.9565           0.0435  PLEARN-v1.2.66   \n",
       "9996  Delta (B.1.617.2-like)          1.0000           0.0000   PANGO-v1.2.66   \n",
       "9997    Alpha (B.1.1.7-like)          0.9565           0.0435   PANGO-v1.2.66   \n",
       "9998    Alpha (B.1.1.7-like)          0.9565           0.0435  PLEARN-v1.2.66   \n",
       "9999  Delta (B.1.617.2-like)          1.0000           0.0000  PLEARN-v1.2.66   \n",
       "\n",
       "     pangolin_version pangoLEARN_version pango_version     status  \\\n",
       "0              3.1.11         2021-08-24       v1.2.66  passed_qc   \n",
       "1              3.1.11         2021-08-24       v1.2.66  passed_qc   \n",
       "2              3.1.11         2021-08-24       v1.2.66  passed_qc   \n",
       "3              3.1.11         2021-08-24       v1.2.66  passed_qc   \n",
       "4              3.1.11         2021-08-24       v1.2.66  passed_qc   \n",
       "...               ...                ...           ...        ...   \n",
       "9995           3.1.11         2021-08-24       v1.2.66  passed_qc   \n",
       "9996           3.1.11         2021-08-24       v1.2.66  passed_qc   \n",
       "9997           3.1.11         2021-08-24       v1.2.66  passed_qc   \n",
       "9998           3.1.11         2021-08-24       v1.2.66  passed_qc   \n",
       "9999           3.1.11         2021-08-24       v1.2.66  passed_qc   \n",
       "\n",
       "                                                   note  \n",
       "0     scorpio call: Alt alleles 22; Ref alleles 1; A...  \n",
       "1     scorpio call: Alt alleles 16; Ref alleles 0; A...  \n",
       "2                       Assigned from designation hash.  \n",
       "3     scorpio call: Alt alleles 14; Ref alleles 0; A...  \n",
       "4     scorpio call: Alt alleles 22; Ref alleles 1; A...  \n",
       "...                                                 ...  \n",
       "9995  scorpio call: Alt alleles 22; Ref alleles 1; A...  \n",
       "9996  scorpio call: Alt alleles 13; Ref alleles 0; A...  \n",
       "9997  scorpio call: Alt alleles 22; Ref alleles 1; A...  \n",
       "9998  scorpio call: Alt alleles 22; Ref alleles 1; A...  \n",
       "9999  scorpio call: Alt alleles 13; Ref alleles 0; A...  \n",
       "\n",
       "[10000 rows x 53 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata_pangolin_df = metadata_df.merge(p_df, left_on='strain', right_on='taxon')\n",
    "metadata_pangolin_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "58d82e66-28b1-4882-8b28-b422eacf3041",
   "metadata": {},
   "source": [
    "# 4. Save joined metadata to file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "9761daae-e204-42d7-84fc-9e49b775fde7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saved metadata and pangolin to input/metadata-subsample-pangolin.tsv\n"
     ]
    }
   ],
   "source": [
    "metadata_pangolin_df.to_csv(output_metadata_pangolin_path, sep='\\t', index=False)\n",
    "\n",
    "print(f'Saved metadata and pangolin to {output_metadata_pangolin_path}')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
